## Chapter 1

#### Dan Stefanica 
##### Chapter 1 Problem 9

Calculated Arrays are   
$Df(a)=\left[5,-1,-3\right]$  
$D^2f(a)=\left[\begin{array}{ccc}
4&-1&0\\
-1&0&3\\
0&3&-6
\end{array}\right]$

$a=\left[\begin{array}{c}1\\-1\\0\end{array}\right]$  

$f(x_1,x_2,x_3)=2x_1^2-x_1x_2+3x_2x_3-x_3^2$  
Prove that  
$f(x)=f(a)+Df(a)(x-a)+\frac{(x-a)^T}{2}D^2f(a)(x-a)$  
$f(a=(1,-1,0))=2*1^2-1*-1+3*-1*0-0^2 =2+1=3$  
$Df(a)(x-a)=\left[\begin{array}{ccc} 5&-1&-3\end{array}\right] * \left[\begin{array}{c} x_1-1\\x_2+1\\x_3\end{array}\right]$   
$\qquad\qquad= 5(x_1-1)-(x_2+1)-3x_3$  
$\qquad\qquad=5x_1-x_2-3x_3-6$  
$\frac{(x-a)^T}{2}D^2f(a)(x-a)=\frac{1}{2}*\left[\begin{array}{ccc} x_1-1&x_2+1&x_3\end{array}\right]\left[\begin{array}{ccc} 
4&-1&0\\
-1&0&3\\
0&3&-2\end{array}\right]\left[\begin{array}{c} x_1-1\\x_2+1\\x_3\end{array}\right]$  
$\qquad\qquad=\frac{1}{2}\left[\begin{array}{ccc} 4x_1-4-x_2-1&-x_1+1+3x_3&3x_2+3-2x_3\end{array}\right]\left[\begin{array}{c} x_1-1\\x_2+1\\x_3\end{array}\right]$  
$\qquad\qquad=\frac{1}{2}\left[\begin{array}{ccc} 4x_1-x_2-5&-x_1+3x_3+1&3x_2-2x_3+3\end{array}\right]\left[\begin{array}{c} x_1-1\\x_2+1\\x_3\end{array}\right]$  
$\qquad\qquad=\frac{1}{2}\left[(4x_1-x_2-5)(x_1-1) + (-x_1+3x_3+1)(x_2+1) + x_3(3x_2-2x_3+3)\right]$  
$\qquad\qquad=\frac{1}{2}\left[(4x_1-x_2-5)(x_1-1) + (-x_1+3x_3+1)(x_2+1) + x_3(3x_2-2x_3+3)\right]$  
$\qquad\qquad=\frac{1}{2}\left[4x_1^2-x_1x_2-x_15    -4x_1+x_2+5  -x_1x_2+3x_2x_3+x_2    -x_1+3x_3+1 + 3x_2x_3-2x_3^2+3x_3\right]$  
$\qquad\qquad=\frac{1}{2}\left[4x_1^2-2x_1x_2-10x_1    +x_2+x_2  +3x_2x_3+ 3x_2x_3    +3x_3+3x_3 -2x_3^2+5+1\right]$  
$\qquad\qquad=\frac{1}{2}\left[4x_1^2-2x_1x_2-10x_1+2x_2-2x_3^2+6x_2x_3+6x_3 +6\right]$  
$\qquad\qquad=2x_1^2-x_1x_2-5x_1+x_2-x_3^2+3x_2x_3+3x_3 +3$  

$f(a)+Df(a)(x-a)+\frac{1}{2}(x-a)^TD^2f(a) (x-a)=$  
$\qquad\qquad 3 +$  
$\qquad\qquad5x_1-x_2-3x_3-6+$  
$\qquad\qquad2x_1^2-x_1x_2-5x_1+x_2-x_3^2+3x_2x_3+3x_3 +3$  
$\rightarrow2x_1^2-x_1x_2+3x_2x_3-x_3^2$  

In [ ]:
#Dan Stefanica
#Chapter1 Exercise Problem 12
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
x=np.array([x for x in range(5,50,5)])
LC30=np.array([max(0,s-30) for s in x])
SC35=np.array([-max(0,s-35) for s in x])
LP20=np.array([max(0,20-s) for s in x])
SP15=np.array([-max(0,15-s) for s in x])
print(x)
print(LC30)
print(SC35)
print(LP20)
print(SP15)
Total=LC30+SC35+LP20+SP15
print(Total)
lines=plt.plot(x,LC30,x,SC35,x,LP20,x,SP15,x,Total)
lines[0].set_label("LC30")
lines[1].set_label("SC35")
lines[2].set_label("LP20")
lines[3].set_label("SP15")
lines[4].set_label("Total")
plt.legend()
plt.xlabel("Spot")
plt.ylabel("PayOff")
plt.show()

In [ ]:
#Dan Stefanica
#Chapter1 Exercise Problem 14
x=np.array([x for x in range(90,170,10)])
LC100=np.array([max(0,s-100) for s in x])
SC120=np.array([-3*max(0,s-120) for s in x])
LC130=np.array([2*max(0,s-130) for s in x])
Total=LC100+SC120+LC130
lines=plt.plot(x,LC100,x,SC120,x,LC130,x,Total)
lines[0].set_label("1-LC100")
lines[1].set_label("3-SC120")
lines[2].set_label("2-LC130")
lines[3].set_label("Total")
plt.legend()
plt.xlabel("Spot")
plt.ylabel("PayOff")
plt.show()

In [ ]:
#Dan Stefanica
#Chapter 1 Exercise Problem 15 
import math
S=40.
div=0.03
T=4/12.
P=2
C=4
r=0.05
#By put call parity

#P+S*exp(-dT)-C=K*exp(-rT)
divDisc=math.exp(-div*T)
rDisc=math.exp(-r*T)
LHS=P+S*divDisc-C
RHS=S*rDisc # as the options are at the money.
print(LHS)
print(RHS)
PVS=S*divDisc

#Buy Stock S +40*divDisc
#Buy       P +2
#Sell      C -4
#TotalCost V +38
V=S*divDisc+P-C
print('Portfolio Cash Cost at t=0 is {}'.format(V))
#Borrow V at risk free rate till T
VT=V/rDisc
print('Loan Value to payoff at t=T is {}'.format(VT))
#S>40
#C pay S-40
#Sell Stock at S
#N=S-(S-40)-VT=40-VT (Net Loan payoff)
#S<=40
#Put pays (40-S)
#Sell Stock S
#N=S+(40-S)-VT=40-VT

N=40-VT
print(N)
print(5.5-42-2.75)


##### Dan Stefanica
###### Chapter 1 Exercise Problem 16

By put Call parity  
**S+P-C=K**    
Cost to Buy the Portfolio  
$K_1$=42+3.25-5=40.25  
Or  
Cash if we short the portfolio  
**C-S-P=-K**  
$K_2$=5.5-42-2.75=-39.25

If we borrow 40.25 to setup the portfolio then we will have to pay 40.25 but portfolio value would be =40. Loss=0.25$

If we invest the proceeds from short portfolio of 39.25 we will have 39.25 at expiry but a cost of 40 to unwind the short positions. Loss=0.75$

In [ ]:
#Dan Stefanica
# Chapter 1 Exercise 18
r=0.08
div=0.03
T=3/12.
disc=math.exp(-r*T)
divDisc=math.exp(-0.03*T)

PVF=12100*disc
PVS=12000*divDisc
print("Val to buy Fut={}".format(PVF))
print("Val to buy Stock={}".format(PVS))

#Short Stock
S=12000*divDisc
#Invest Proceeds in Risk Free
VT=S/disc
#Buy Futures at 12100

#At maturity
#Pay 12100 to buy the index in the futures contract
N=VT-12100
print(N)



## Chapter 2

In [ ]:
#Dan Stefanica
#Chapter 2 Exercise problem 7

r=[0.05,0.0525,0.0535,0.055]
t=[0.5,1.0,1.5,2.0]
c=[2.5,2.5,2.5,102.5]
disc=[math.exp(-r[i]*t[i]) for i in range(len(r))]
P=sum(np.array(c)*np.array(disc))
print("Bond Price is {}".format(P))

In [ ]:
#Dan Stefanica
#Chapter 2 Exercise problem 8
PV_Of_FV=100*disc[-1]
PV_Of_Coupons=100-PV_Of_FV
Semi_Annual_Par_Coupon=PV_Of_Coupons/sum(disc)
print("Par Coupon or par yield for the bond is {}".format(Semi_Annual_Par_Coupon*2))

In [ ]:
#Dan Stefanica
#Chapter 2 Excersice problem 10
Cpn=0.06
T=30/12.
Y=0.09
CF=np.full((5),3)
CF[-1]+=100
t=np.arange(0.5,30/12+0.5,0.5)
disc=np.array([math.exp(-Y*tp) for tp in t])
BondPrice=sum(CF*disc)
print("Bond Price is {}".format(BondPrice))
Duration=sum(CF*t*disc)/BondPrice
print("Bond Duration is {}".format(Duration))
Convexity=sum(CF*t*t*disc)/BondPrice
print("Bond Convexity is {}".format(Convexity))

In [ ]:
#Dan Stefanica
#Chapter 2 Exercise problem 11
Cpn=0.08
Freq=4
T=14/12.
Y=0.07
t=np.arange(T,0,-1./Freq)
t=t[::-1]
print(t)
CF=np.full(len(t),Cpn*100/4)
CF[-1]+=100
print(CF)
disc=np.array([math.exp(-Y*tp) for tp in t])
print(disc)
BondPrice=sum(CF*disc)
print("Bond Price is {}".format(BondPrice))
Duration=sum(CF*t*disc)/BondPrice
print("Bond Duration is {}".format(Duration))
Convexity=sum(CF*t*t*disc)/BondPrice
print("Bond Convexity is {}".format(Convexity))

In [ ]:
#Dan Stefanica
#Chapter 2 Exercise problem 12

r=lambda t:0.05+0.01*math.log(1+t/2.)

T=2
Freq=2
Cpn=0.08
t=np.arange(T,0,-1./Freq)
t=t[::-1]
disc=np.array([math.exp(-r(tp)*tp) for tp in t])
CF=np.full(len(t),Cpn*100/Freq)
CF[-1]+=100
print(CF)
print(t)
print(disc)
BondPrice=sum(CF*disc)
print("Bond Price is {}".format(BondPrice))
#For Calculatin Duration and Convexity, we have to do it with yield and not zero rates
#Use Newton Raphson method to compute Yield of the Bond.

#The below methodology is wrong as we are using zero rates to compute the D and C
Duration=sum(CF*t*disc)/BondPrice
print("Bond Duration is {}".format(Duration))
Convexity=sum(CF*t*t*disc)/BondPrice
print("Bond Convexity is {}".format(Convexity))

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
y=lambda x: (x-3.)*(x-2.)

x=np.arange(-10,10,1)
plt.plot(x,[y(xp) for xp in x])
plt.show()

## Chapter 3

Example Workout

Use the Black-Scholes formula to price a six months European call option with strike 40, on an underlying asset with spot price 42 and volatility 30%, which pays dividends continuously, with dividend rate 3%. Assume that interest rates are constant at 5%.

In [ ]:
import math
from scipy.stats import norm
K=40
S=42
V=0.3
q=0.03
r=0.05
t=0.5
#C=Sexp(-qt)N(d1)-Kexp(-rt)N(d2)
#d1=log(S/K)+(r-q+V^2/2)t / V sqrt(t)
srt=V*t**0.5
d1=(math.log(S/K)+(r-q+V**2/2)*t)/srt
d2=d1-srt
C=S*math.exp(-q*t)*norm.cdf(d1)-K*math.exp(-r*t)*norm.cdf(d2)
print("Call Option price is {}".format(C))

Dan Stefanica
Chapter 3 Exercise Problem 6

6) Use the Black-Scholes formula to price both a put and a call option with strike K = 45 expiring in six months on an underlying asset with spot price 50 and volatility 20% paying dividends continuously at 2%, if interest rates are constant at 6%.

In [ ]:
#Dan Stefanica
#Chapter 3 Exercise Problem 6
K=45
S=50
V=0.2
q=0.02
r=0.06
t=0.5
#C=Sexp(-qt)N(d1)-Kexp(-rt)N(d2)
#d1=log(S/K)+(r-q+V^2/2)t / V sqrt(t)
srt=V*t**0.5
d1=(math.log(S/K)+(r-q+V**2/2)*t)/srt
d2=d1-srt
C=S*math.exp(-q*t)*norm.cdf(d1)-K*math.exp(-r*t)*norm.cdf(d2)
print("Call Option price is {}".format(C))
P=K*math.exp(-r*t)-S*math.exp(-q*t)+C
print("Put Option price is {}".format(P))

Dan Stefanica
Chapter 3 Exercise Problem 3

3. Recall the stock evolution model from section 3.1: over each of three consecutive time intervals of length T = 1/12, a stock with spot price So = 40 at time t = 0 will either go up by a factor u = 1.05 with probability p = 0.6, or down by a factor d = 0.96 with probability 1 - p = 0.4. Compute the expected value and the variance of the stock price at time T = 3T, i.e., compute E[ST] and var(ST).

In [ ]:
from math import factorial as fact
import numpy as np
C=lambda n,r:fact(n)/fact(r)/fact(n-r)
s=40
u=1.05
d=0.96
pu=0.6
pd=0.4
n=3
S=np.array([s*u**(n-i)*d**i for i in range(n+1)])
P=np.array([C(n,i)*pu**(n-i)*pd**i for i in range(n+1)])
ES=sum(S*P)
print("Expected Stock Price at 3T is {}".format(ES))
ES2=sum(S*S*P)
VS=ES2-ES*ES
print("Variance of Stock Price at 3T is {}".format(VS))


17) You are long 1000 call options with strike 90 and three months to ma- turity. Assume that the underlying asset has a lognormal distribution
with drift $\mu$ = 0.08 and volatility $\sigma$= 0.2, and that the spot price o
the asset is 92. The risk-free rate is r = 0.05. What Delta-hedging position do you need to take? (Note that, for Delta-hedging purposes, it is not necessary to know the drift $\mu$ of the underlying asset, since $\Delta(C)$ does not depend on $\mu$; cf. (3.66).)

In [ ]:
#Dan Stefanica
#Chapter3 Exercise Problem 17
d1=lambda S,K,r,q,v,t:(math.log(S/K)+(r-q+(v**2)/2)*t)/(v*(t**0.5))
d2=lambda d_1,v,t:d_1-v*(t**0.5)
S=92.
K=90.
r=0.05
q=0
v=0.2
t=3/12.

#Delta(CallOption)=N(d1) 
#                  as q=0
DeltaC=norm.cdf(d1(S,K,r,q,v,t))
print("Number of stocks to short for hedging the 1000 call options is {}".format(DeltaC*1000))

18) You buy 1000 six months ATM Call options on a non-dividend-paying asset with spot price 100, following a lognormal process with volatility 30%. Assume the interest rates are constant at 5%.  
(i) How much money do you pay for the options?  
(ii) What Delta-hedging position do you have to take?  
(iii) On the next trading day, the asset opens at 102. What is the value of your position (the option and shares position)?  
(iv) Had you not Delta-hedged, how much would you have lost due to the increase in the price of the asset?  

In [ ]:
#Dan Stefanica
#Chapter3 Exercise Problem 18
d1=lambda S,K,r,q,v,t:(math.log(S/K)+(r-q+(v**2)/2)*t)/v/t**0.5
d2=lambda d_1,v,t:d_1-v*t**0.5
q=0
S=100
K=100
V=0.3
r=0.05
t=0.5
#C=SN(d1)-Kexp(-rt)N(d2)
d_1=d1(S,K,r,q,V,t)
d_2=d2(d_1,V,t)
C=S*norm.cdf(d_1)-K*math.exp(-r*t)*norm.cdf(d_2)
V1=C*1000
print("Price to pay for 1000 ATM Call Options is {:.2f}".format(V1))
Hedge=math.ceil(norm.cdf(d_1)*1000)
Hedge1=-Hedge*S
print("Stocks needed to short for hedging={:.2f}".format(Hedge))
#Next day when spot moves to 102
S=98
t=125/252
d_1=d1(S,K,r,q,V,t)
d_2=d2(d_1,V,t)
C=S*norm.cdf(d_1)-K*math.exp(-r*t)*norm.cdf(d_2)
V2=C*1000
print("New Value of 1000 ATM Call Options is {:.2f}".format(V2))
Hedge2=-Hedge*S
print("New Value of portfolio={:.2f}".format(V2+Hedge2))
print("Change in Value of Options only={:.2f}".format(V2-V1))
print("Total Change in value of portfolio={:.2f}".format(V2+Hedge2-V1-Hedge1))


Dan Stefanica
Chapter 3 Exercise Problem 8

8) Use formula (3.74), i.e., $p(C) = K(T - t)e^{-r(T-t)}N(d2)$, and the Put- Call parity to obtain the formula (3.75) for p(P), i.e.,
$p(P) = - K(T - t)e^{-r(T-t)}N(-d2)$.

**Use put call parity for calculating rho of put option**
8. Use formula (3.74), i.e., $\rho(C) = K(T - t)e^{-r(T-t)}N(d_2)$, and the Put- Call parity to obtain the formula (3.75) for $\rho(P)$, i.e.,
$\rho(P) = - K(T - t)e^{-r(T-t)}N(-d_2)$.  
$P+Se^{-q(T-t)}-C=Ke^{-r(T-t)}$  
$\rho(P)-\rho(C)=-(T-t)Ke^{-r(T-t)}$  
$\rho(P)=\rho(C)-(T-t)Ke^{-r(T-t)}$  
$\rho(P)=K(T - t)e^{-r(T-t)}N(d_2)-K(T-t)e^{-r(T-t)}$  
$\rho(P)=K(T - t)e^{-r(T-t)}(N(d_2)-1)$  
$\rho(P)=-K(T - t)e^{-r(T-t)}N(-d_2)$  


Dan Stefanica  
Chapter 3 Exercise Problem 13

13) Compute the Gamma of ATM call options with maturities of fifteen days, three months, and one year, respectively, on a non-dividend- paying underlying asset with spot price 50 and volatility 30%. Assume that interest rates are constant at 5%. What can you infer about the hedging of ATM options with different maturities?

In [ ]:
#Dan Stefanica
#Chapter 3 Exercise Problem 13 (Answer does not match solution given)
S=50
K=50
V=0.3
q=0
r=0.05
t=[1/24.,3/12.,1.]

#Delta = exp(-qt)N(d1)
#Gamma = exp(-d1^2 /2)/[SV sqrt(2*pi*t)]
#d1=[log(S/K)+(r+v^2 /2)t]/[v sqrt(t)]
#d1=(r+v^2 /2)sqrt(t)/v

D1=[(r+V**2/2)*tp**0.5/V for tp in t]
Gamma=[math.exp(-D1[i]**2/2)/(S*V*(2*math.pi*t[i])**0.5) for i in range(len(t))]
print(Gamma)

## Chapter 4


8) Consider a put option with strike 55 and maturity 4 months on a non- dividend paying asset with spot price 60 which follows a lognormal model with drift $\mu$ = 0.1 and volatility $\sigma$ = 0.3. Assume that the risk-free rate is constant equal to 0.05.  
(i) Find the probability that the put will expire in the money.  
(ii) Find the risk-neutral probability that the put will expire in the money.  
(iii) Compute N(-d2).

In [ ]:
S=60
K=55
T=4/12.
V=0.3
r=0.05
q=0
u=0.1

d_1=d1(S,K,u,q,V,T)
d_2=d2(d_1,V,T)
print("Probability of put being in the money is {}".format(norm.cdf(-d_2)))
#The above is not risk free as the Stock has it's drift of u different from the risk free rate r.

d_1=d1(S,K,r,q,V,T)
d_2=d2(d_1,V,T)
print("Risk Free Probability of put being in the money is {}".format(norm.cdf(-d_2)))


9) (i) Consider an at-the-money call on a non-dividend paying asset;  
assume the Black-Scholes framework. Show that the Delta of the option is always greater than 0.5.  
(ii) If the underlying pays dividends at the continuous rate q, when is the Delta of an at-the-money call less than 0.57  
Note: For most cases, the Delta of an at-the-money call option is close to 0.5.

$Delta=N(d_1)$  
Now we have to show that $N(d_1)>0.5$ for a non dividend paying stock.  
$\implies d_1>0$  
$\implies \frac{log(S/K)+(r+\sigma^2/2)T}{\sigma\sqrt T}>0$  
$\implies \frac{log(S/K)+(r+\sigma^2/2)T}{\sigma\sqrt T}>0\quad$As the Option is ATM $log(S/K)=0$  
$\implies \frac{(r+\sigma^2/2)\sqrt T}{\sigma}>0\quad$  
As all the factors in the equation are positive the above value should be >0.  

Delta of Call Option would be <0.5 if  
$\implies r-q+\sigma^2/2<0\quad$  
$\implies q>r+\sigma^2/2\quad$  





10) Use risk-neutral pricing to price a supershare, i.e., and option that pays $(max(S(T) - K,0))^2)$ at the maturity of the option. In other words, compute
$V(O) = e^{-rT} E_{RN}[(max(S(T) - K,0))^2]$,
where the expected value is computed with respect to the risk-neutral distribution of the price S(T) of the underlying asset at maturity T, which is assumed to follow a lognormal process with drift r and volatility $\sigma$. Assume that the underlying asset pays no dividends, i.e., q = O.

$S_T=S_0e^{(r-\sigma^2/2)T+\sigma\sqrt T Z}$  
$P(S_T>K)~N(d_2)$  
$V_0=\frac{e^{-rT}}{\sqrt{2\pi}}\int_{-\infty}^{d2} (S_0e^{(r-\sigma^2/2)T+\sigma\sqrt T x}-K)^2e^{-\frac{x^2}{2}}dx$  
$(S_T-K)^2=S_T^2-2S_TK+K^2$  

$E[K^2]=\frac{K^2e^{-rT}}{\sqrt{2\pi}}\int_{-\infty}^{d2}e^{-x^2/2}dx = K^2e^{-rT}N(d_2)$  
$E[2S_TK]=\frac{KS_0e^{-rT}}{\sqrt{2\pi}}\int_{-d_2}^{\infty}e^{(r-\sigma^2/2)T + \sigma\sqrt T x - \frac{x^2}{2}}dx$ 
$\qquad=\frac{KS_0}{\sqrt{2\pi}}\int_{-d_2}^{\infty}e^{\frac{-\sigma^2T+2\sigma\sqrt T x-x^2}{2}}$$dx$  
$\qquad=\frac{KS_0}{\sqrt{2\pi}}\int_{-d_2}^{\infty}e^{\frac{-\sigma^2T+2\sigma\sqrt T x-x^2}{2}}dx$  
$\qquad=\frac{KS_0}{\sqrt{2\pi}}\int_{-d_2}^{\infty}e^{-\frac{(x-\sigma\sqrt T)^2}{2}}dx$  
$\qquad=\frac{KS_0}{\sqrt{2\pi}}\int_{-d_2-\sigma\sqrt T}^{\infty}e^{-\frac{t^2}{2}}dt$  
$\qquad=\frac{KS_0}{\sqrt{2\pi}}\int_{-d_1}^{\infty}e^{-\frac{t^2}{2}}dt$  
$\qquad=KS_0N(d_1)$    
$E[S_T^2]=\frac{S_0^2e^{-rT}}{\sqrt{2\pi}}\int_{-d_2}^{\infty} e^{(2r-\sigma^2)T+2\sigma\sqrt T x -\frac{x^2}{2}}dx$  
$\qquad=\frac{S_0^2e^{rT}}{\sqrt{2\pi}}\int_{-d_2}^{\infty} e^{-\sigma^2T+2\sigma\sqrt T x -\frac{x^2}{2}}dx$  
$\qquad=\frac{S_0^2e^{rT}}{\sqrt{2\pi}}\int_{-d_2}^{\infty} e^{\frac{-4\sigma^2T+4\sigma\sqrt T x -x^2}{2} + \sigma^2T}dx$  
$\qquad=\frac{S_0^2e^{rT+ \sigma^2T}}{\sqrt{2\pi}}\int_{-d_2}^{\infty} e^{-\frac{(x-2\sigma\sqrt T)^2}{2} }dx$  
$\qquad=\frac{S_0^2e^{rT+ \sigma^2T}}{\sqrt{2\pi}}\int_{-d_2-2\sigma\sqrt T}^{\infty} e^{-\frac{t^2}{2} }dt$  
$\qquad=\frac{S_0^2e^{rT+ \sigma^2T}}{\sqrt{2\pi}}\int_{-d_1-\sigma\sqrt T}^{\infty} e^{-\frac{t^2}{2} }dt$  
$\qquad=S_0^2e^{rT+ \sigma^2T}N(d_1+\sigma\sqrt T)$  
$V_0=K^2e^{-rT}N(d_2)-KS_0N(d_1)+S_0^2e^{(r+ \sigma^2)T}N(d_1+\sigma\sqrt T)$  

11) If the price of an asset follows a normal process, i.e., $dS = \mu dt +\sigma dX$, then  
Assume that the risk free rate is constant and equal to r.  
(i) Use risk neutrality to find the value of a call option with strike K
and maturity T, i.e., compute
$C(0) = e^{ - r T} E_{RN}[max(S(T) - K,0)]$,
where the expected value is computed with respect to S(T) given by   
$S(T) = S(O) +rT +r\sqrt T Z$  
(ii) Use the Put-Call parity to find the value of a put option with strike K and maturity T, if the underlying asset follows a normal process as above.

$P(S_T>K)$  
$P(S_0+rT+\sigma\sqrt T Z>K)$  
$P(Z>\frac{K-S_0-rT}{\sigma\sqrt T})$  
Let $\frac{K-S_0-rT}{\sigma\sqrt T}=k$  
$C_0=\frac{e^{-rT}}{\sqrt{2\pi}}\int_k^\infty (S_0+rT+\sigma\sqrt T x-K)e^{-\frac{x^2}{2}}dx$  
$\qquad=(S_0+rT-K)e^{-rT}N(-k)+ \frac{e^{-rT}}{\sqrt{2\pi}}\int_k^\infty (\sigma\sqrt T x)e^{-\frac{x^2}{2}}dx$  
$\qquad=(S_0+rT-K)e^{-rT}N(-k)+ \frac{e^{-rT}\sigma\sqrt T}{\sqrt{2\pi}}\int_k^\infty  xe^{-\frac{x^2}{2}}dx$  
$\qquad=(S_0+rT-K)e^{-rT}N(-k)+ \frac{e^{-rT}\sigma\sqrt T}{\sqrt{2\pi}}\int_k^\infty  d(-e^{-\frac{x^2}{2}})$  
$\qquad=(S_0+rT-K)e^{-rT}N(-k)+ \frac{e^{-rT}\sigma\sqrt T}{\sqrt{2\pi}}\left[-e^{-\frac{x^2}{2}}\right]_k^\infty$  
$\qquad=(S_0+rT-K)e^{-rT}N(-k)+ \frac{e^{-rT}\sigma\sqrt T}{\sqrt{2\pi}}e^{-\frac{k^2}{2}}$  
$\qquad=(S_0+rT-K)e^{-rT}N(-k)+ \frac{\sigma\sqrt T}{\sqrt{2\pi}}e^{-rT-\frac{k^2}{2}}$  


## Chapter 5

4) Let  
$\qquad\qquad T(x) =\sum_{k=1}^\infty\frac{(-1)^{k+1} x^k}{k}$  
be the Taylor series expansion of $f(x) = ln (1 + x)$. In section 5.3.1, we showed that T(x) = f(x) if $|x|\leq \frac{1}{2}$. In this exercise, we show that T(x) = f(x) for all x such that |x| < 1.  
Let Pn (x) be the Taylor polynomial of degree n corresponding to f (x).  
Since $T(x) = \operatorname{limn}_{n\rightarrow\infty}Pn(x)$, it follows that f(x) = T(x) for all |x| < 1 iff  
$lim_{n\rightarrow\infty} |f(x) - Pn(x)| = 0,\, \forall |x| < 1.$  
  
1.  Use (5.58) and the integral formula (5.4) for the Taylor approximation error to show that, for any x,  
$f(x)-P_n(x)=\int_0^x \frac{ (-1)^{n+2} (x-t)^n}{ (1+t)^{n+l}} dt$.  
  
2.  Show that, for any $0 \leq x < 1$,  
$|f(x) - P_n(x)| \leq \int_0^x \left(\frac{x-t}{1+t}\right)^n \frac{1}{1+t} dt \leq x^nln(l +x)$. (5.100)  
Use (5.100) to prove that (5.99) holds for all x such that $0 \leq x < 1$.  
  
3. Assume that $-1 < x \leq 0$. Let s = -x. Show that  
$|f(x) - P_n(x)| = \int_0^s\frac{(s-z)^n}{(1-z)^{n+1}}dz$.  
  
Note that $\frac{s-z}{1-z} \leq s$, for all $0 \leq z \leq s < 1$, and obtain that  
$\quad|f(x) - Pn(x)| \leq s^n|ln(l - s)| = (-x)^n|ln(l + x)|$·  

Conclude that (5.99) holds true for all x such that $-1 < x \leq 0$.

$f(x)=ln(1+x)$


n|$f^{(n)}(x)$|$f^{(n)}(0)$
---|---|---
0|$ln(1+x)$|$0$
1|$\frac{1}{1+x}$|$1$
2|$\frac{-1^11!}{(1+x)^2}$|$-1^1$
3|$\frac{-1^22!}{(1+x)^3}$|$-1^22!$
4|$\frac{-1^33!}{(1+x)^4}$|$-1^33!$
.|.|.
.|.|.
.|.|.
k|$\frac{-1^{(k+1)}(k-1)!}{(1+x)^k}$|$-1^{(k+1)}(k-1)!$  

For Radius of convergence:-  
$R=\frac{1}{\operatorname{limsup}_{k\rightarrow\infty}|a_k|^{\frac{1}{k}}}$  
$\operatorname{limsup}_{k\rightarrow\infty}|a_k|^{\frac{1}{k}}  
=\operatorname{limsup}_{k\rightarrow\infty}|\frac{-1^{(k+1)}(k-1)!}{k!}|^{\frac{1}{k}}=Lt_{k\rightarrow\infty}(\frac{1}{k})^{\frac{1}{k}}=1$  
$R=1$  

Now we need to find the range within [-1,1] where T(x) would converge to f(x). Note that as of now we only know that T(x) converges in (-1,1) but we do not know if it converges to a proper f(x).

For limiting error to zero(using Integral error):-  
$Lt_{n\rightarrow\infty}|f(x)-P_n(x)|=Lt_{n\rightarrow\infty}\left|\int_0^x \frac{(x-t)^n}{n!}f^{(n+1)}(t)dt\right|=0$  
$\left|\int_0^x \frac{(x-t)^n}{n!}f^{(n+1)}(t)dt\right|\\=\left|\int_0^x \frac{(x-t)^n}{n!}\frac{-1^{n+2}n!}{(1+t)^{n+1}}\,dt\right|\\=\left|\int_0^x \frac{(x-t)^n}{(1+t)^{n+1}}\,dt\right|\\=\left|\int_0^x \left(\frac{x-t}{1+t}\right)^n\frac{1}{1+t}\,dt\right|$

**Using a slightly different method than suggested in the question**  
Review $\frac{x-t}{1+t}$  

if $-1<x\leq t\leq 0$  

$x<\frac{x-t}{1+t}\implies x+tx<x-t$ which is true  
$\implies |x|>|\frac{x-t}{1+t}|$  
$\implies |x|^n>|\frac{x-t}{1+t}|^n$  

if $0\leq t\leq x<1$    

$x>\frac{x-t}{1+t}\implies x+tx>x-t$ which is true  
$\implies |x|>|\frac{x-t}{1+t}|$  
$\implies |x|^n>|\frac{x-t}{1+t}|^n$  

$\left|\int_0^x \left(\frac{x-t}{1+t}\right)^n\frac{1}{1+t}\,dt\right| \leq \left|\int_0^x \left|\frac{x-t}{1+t}\right|^n\frac{1}{1+t}\,dt\right|\leq\left|\int_0^x |x|^n\frac{1}{1+t}\,dt\right|=|x|^n\left|\int_0^x \frac{1}{1+t}\,dt\right|=|x|^n|ln(1+t)|_0^x$


$Lt_{n\rightarrow\infty}|f(x)-P_n(x)|\leq Lt_{n\rightarrow\infty}|x|^n|ln(1+x)|$  

$Lt_{n\rightarrow\infty}|x|^n|ln(1+x)|=0\quad \forall\, x\,\epsilon\,(-1,1)$

$Lt_{n\rightarrow\infty}|f(x)-P_n(x)|=0 \quad \forall\, x\,\epsilon\,(-1,1)$  
Thus T(x)=f(x) in the range (-1,1).  

Exercise Problem 5

$A=\frac{1}{2}(e^{-r\delta t}+e^{(r+\sigma^2)\delta t})$  

$e^{-r\delta t}=1-r\delta t+O(\delta t^2)$  
$e^{(r+\sigma^2)\delta t}  =1+(r+\sigma^2)\delta t+ O(\delta t^2)$  
$A=\frac{1}{2}(1-r\delta t+1+(r+\sigma^2)\delta t) +O(\delta t^2)$   
$A=1+\frac{\sigma^2\delta t}{2} + O(\delta t^2)$  
$\sqrt{A^2-1}\\=\sqrt{(1+\frac{\sigma^2\delta t}{2})^2-1}\\=\sqrt{1+\frac{\sigma^4\delta t^2}{4}+2\frac{\sigma^2\delta t}{2}-1}\\=\sqrt{\frac{\sigma^4\delta t^2}{4}+\sigma^2\delta t}\\=\sigma\sqrt\delta t\sqrt{1+\frac{\sigma^2\delta t}{4}}$  
$\sqrt{1+x}=1+\frac{x}{2}+O(x^3)$  
$\sqrt{A^2-1}=\sigma\sqrt\delta t(1+\frac{\sigma^2\delta t}{8}+O(\delta t^3))=\sigma\sqrt\delta t +O(\delta t^\frac{3}{2}) + O(\delta t^\frac{7}{2})=\sigma\sqrt\delta t +O(\delta t^\frac{3}{2})$  
$u= A+\sqrt{A^2-1}\\=1+\frac{\sigma^2\delta t}{2}+ O(\delta t^2) + \sigma\sqrt\delta t+ O(\delta t^\frac{3}{2})\\=1 + \sigma\sqrt\delta t+\frac{\sigma^2\delta t}{2} +O(\delta t^\frac{3}{2})$  

$e^{\sigma\sqrt\delta t}=1+\sigma\sqrt\delta t+\frac{\sigma^2\delta t}{2}+O((\sigma\sqrt\delta)^3)\\=1+\sigma\sqrt\delta t+\frac{\sigma^2\delta t}{2}+O(\delta t^\frac{3}{2})$  
$e^{-\sigma\sqrt\delta t}\\=1-\sigma\sqrt\delta t+\frac{\sigma^2\delta t}{2}+O(\delta t^\frac{3}{2})$  

$u= A+\sqrt{A^2-1}=1 + \sigma\sqrt\delta t+\frac{\sigma^2\delta t}{2} +O(\delta t^\frac{3}{2})=e^{\sigma\sqrt\delta t}$  
$d= A-\sqrt{A^2-1}=1 - \sigma\sqrt\delta t+\frac{\sigma^2\delta t}{2} +O(\delta t^\frac{3}{2})=e^{-\sigma\sqrt\delta t}$  

**Part 2**  
$p=\frac{e^{r\delta t} - d}{u-d}\\=\frac{e^{r\delta t} - e^{-\sigma\sqrt\delta t}}{e^{\sigma\sqrt\delta t}-e^{-\sigma\sqrt\delta t}}\\=\frac{1+r\delta t+O(\delta t^2) - (1-\sigma\sqrt\delta t + \frac{1}{2}\sigma^2\delta t+O(\delta t^\frac{3}{2})}{1+\sigma\sqrt\delta t +\frac{1}{2}\sigma^2\delta t +O(\delta t^\frac{3}{2})-(1-\sigma\sqrt\delta t +\frac{1}{2}\sigma^2\delta t +O(\delta t^\frac{3}{2})}\\=\frac{r\delta t+\sigma\sqrt\delta t - \frac{1}{2}\sigma^2\delta t+O(\delta t^\frac{3}{2})}{2\sigma\sqrt\delta t + O(\delta t^\frac{3}{2})}$  
$=\frac{\sigma+(r - \frac{1}{2}\sigma^2)\sqrt\delta t+O(\delta t)}{2\sigma + O(\delta t)}$  
$=\frac{1}{2}+\frac{1}{2\sigma}(r - \frac{1}{2}\sigma^2)\sqrt\delta t$  


Exercise Problem 6

$P=C=S(N(d_1)-N(d_2))$  
$\quad=S(N(\frac{\sigma\sqrt t}{2})-N(-\frac{\sigma\sqrt t}{2}))$  
$\quad=S(N(\frac{\sigma\sqrt t}{2})-(1-N(\frac{\sigma\sqrt t}{2})))$  
$\quad=S(2N(\frac{\sigma\sqrt t}{2})-1)$  
$N(x)=\int_{-\infty}^xe^{-\frac{t^2}{2}}dt$  
$N(x)=N(0)+xN'(0)+O(x^2)$  
$N(0)=1$  
$N'(t)=\frac{1}{\sqrt{2\pi}}\frac{d}{dt}\int_{-\infty}^te^{-\frac{x^2}{2}}\,dx\\=\frac{e^{-\frac{t^2}{2}}}{\sqrt{2\pi}}$  
$N'(0)=\frac{1}{\sqrt{2\pi}}$  
$N(x)=N(0)+xN'(0)+O(x^2)\\=\frac{1}{2}+\frac{x}{\sqrt{2\pi}}+O(x^2)$  
$P=S(2N(\frac{\sigma\sqrt t}{2})-1)\\=S(2(\frac{1}{2}+\frac{\sigma\sqrt t}{2\sqrt{2\pi}}+O(t))-1)\\=S(\sigma\sqrt\frac {t}{2\pi} + O(t))\\=\sigma S\sqrt\frac {t}{2\pi} + O(t)$  
$P=0.25*60*\sqrt\frac{1}{2\pi}$

In [ ]:
P1=0.25*60/(2*math.pi)**0.5
print("ATM put price for r=q=0 and S=T is {:.6f}".format(P1))

v=0.25
S=60
t=1

#From Black Scholes
#P=S(N(d1)-N(d2))
#d1=v*sqrt(t)/2
#d2=-v*sqrt(t)/2
Nd1=norm.cdf(0.25/2)
Nd2=1-Nd1
P2=60*(Nd1-Nd2)
print("Black Scholes ATM put price for r=q=0 and S=T is {:.6f}".format(P2))
print("Error from approx is {:.2f}%".format((P2-P1)/P1*100))


9) A five year bond worth 101 has duration 1.5 years and convexity equal to 2.5. Use both formula (2.59), which does not include any convexity adjustment, and formula (5.96) to find the price of the bond if the yield increases by ten basis points (i.e., 0.001), fifty basis points, one percent, and two percent, respectively.

In [ ]:
P=101
D=1.5
C=2.5
dr=[10,50,100,200]

#D=-1/P dP/dr
#dP=-d*dr*P
#C=1/P d2(P)/d(r2)
#P+dP=P-dr*D*P+(dr)^2/2C*P

P1=np.array([P-D*drp*P/10000. for drp in dr])
P2=np.array([P-D*drp*P/10000. + drp**2/2*C*P/10**8 for drp in dr])
print("Duration Prices:{}".format(P1))
print("Convexity Adj Prices:{}".format(P2))
print("Error Percentages:{}".format((P2-P1)/P2*100))

$\sqrt{2\pi}=\sqrt{2*3.142}=\sqrt{6.1284}=\sqrt{4*1.5321}=2*\sqrt{1+0.5321}$   
$\sqrt{1+x}=1+x/2-x^2/8+O(x^3)$
$\sqrt{2\pi}=2*(1+0.5321/2)=2*1.26605=2.53210$

In [ ]:
print(1/2.5321)

Solve ODE for interval $x\epsilon [0,1]$  
$(1 + x)y"(x) + xy'(x) - 2y(x) = 0$  
$y(0) = 1; y(1) = 1$.  

$y"_k(x)=\frac{y_{k-1}-2y_k+y_{k+1}}{h^2} + O(h^2)$  
$y'_k(x)=\frac{y_{k+1}-y_{k-1}}{2h} + O(h^2)$    
$(1 + kh)\frac{y_{k-1}-2y_k+y_{k+1}}{h^2} + kh\frac{y_{k+1}-y_{k-1}}{2h} - 2y_k=0$  
$\frac{1+kh-\frac{kh^2}{2}}{h^2}y_{k-1} + -2\frac{1+kh+h^2}{h^2}y_k + \frac{1+kh+\frac{kh^2}{2}}{h^2}y_{k+1}=0$  
$(1+kh-\frac{kh^2}{2})y_{k-1} + -2(1+kh+h^2)y_k + (1+kh+\frac{kh^2}{2})y_{k+1}=0$  

For k=1  
$(1+h-\frac{h^2}{2}) + -2(1+h+h^2)y_1 + (1+h+\frac{h^2}{2})y_2=0$  
$ -2(1+h+h^2)y_1 + (1+h+\frac{h^2}{2})y_2=-(1+h-\frac{h^2}{2})$

For k=n-1  
$(1+(n-1)h-\frac{(n-1)h^2}{2})y_{n-2} + -2(1+(n-1)h+h^2)y_{n-1} + (1+(n-1)h+\frac{(n-1)h^2}{2})y_n=0$  
$(1+(n-1)h-\frac{(n-1)h^2}{2})y_{n-2} + -2(1+(n-1)h+h^2)y_{n-1} =-(1+(n-1)h+\frac{(n-1)h^2}{2})$  
$(1+(n-1)h-\frac{(n-1)h^2}{2})y_{n-2} + -2(1+(n-1)h+h^2)y_{n-1} =-(1-h+\frac{(1-h)h}{2})$  

In [ ]:
import pandas as pd
iters=8
results=np.full((iters,7),0.)
for k in range(3,3+iters):
    n=2**k
    h=1./n
    A=np.full((n-1,n-1),0.)
    b=np.full((n-1),0.)
    b[0]=(1+h-h**2/2)
    b[-1]=(1+(n-1)*h+(n-1)*h**2/2)
    for i in range(1,len(b)+1):
        A[i-1,i-1]=+2*(1+i*h+h**2)
        if i-1>0:
            A[i-1,i-2]=-(1+i*h-i*h**2/2.)
        if i<len(b):
            A[i-1,i]=-(1+i*h+i*h**2/2.)
    y=np.matmul(np.linalg.inv(A),b)
    #print(y)
    for i in range(1,8):
        results[k-3,i-1]=y[n//8*i-1]
df=pd.DataFrame(results,index=[2**k for k in range(3,3+iters)],columns=["y"+str(k)+"n/8" for k in range(1,8)])
print(df)

### Chapter 7

1. For t = 0 and q = 0, the formula (3.68) for the Gamma of a plain vanilla European call option reduces to  
$\Gamma_C=\Gamma_P=\frac{e^{-q(T-t)}}{S\sigma\sqrt{T-t}}e^{-\frac{d_1^2}{2}}$  
Show that, as a function of S > 0, the Gamma of the call option is first increasing until it reaches a maximum point and then decreases. Also, show that  
$lim_{S\rightarrow0}\Gamma(S)=0$ and $lim_{S\rightarrow\infty}\Gamma(S)=0$  

Ans:  
$\Gamma_C=\Gamma_P=\frac{e^{-q(T-t)}}{S\sigma\sqrt{T-t}}e^{-\frac{d_1^2}{2}}$  
$\Gamma_C=\frac{1}{S\sigma\sqrt{T}}e^{-\frac{d_1^2}{2}}$  


7. The price of a non-dividend-paying asset is lognormally distributed. Assume that the spot price is 40, the volatility is 30%, and the interest rates are constant at 5%. Fill in the Black-Scholes values of the OTM put options on the asset in the table below:  

Option Type| Strike |Maturity|Value      
---|---|---|---
Put| 45| 6 months||
Put| 45|3 months|| 
Put| 48| 6 months||
Put| 48|3 months|| 
Put| 51|6 months|| 
Put| 51|3 months||


For which of these options is the intrinsic value max(K - S,O) larger than the price of the option (in which case the corresponding American put is guaranteed to be worth more than the European put)?

In [ ]:
from scipy.stats import norm
import numpy as np
import math
import pandas as pd
d1=lambda S,K,r,q,v,T:(math.log(S/K)+ (r-q+v**2/2)*T)/v/T**0.5
d2=lambda S,K,r,q,v,T:(math.log(S/K)+ (r-q-v**2/2)*T)/v/T**0.5
P=lambda S,K,r,q,v,T:K*math.exp(-r*T)*norm.cdf(-d2(S,K,r,q,v,T)) -S*math.exp(-q*T)*norm.cdf(-d1(S,K,r,q,v,T))
K=[45,45,48,48,51,51]
T=[6/12,3/12,6/12,3/12,6/12,3/12]
v=0.3
S=40
r=0.05
q=0
POs=np.array([P(S,K[i],r,q,v,T[i]) for i in range(len(K))])
IntrinsicVs=np.array([max(K[i]-S,0) for i in range(len(K))])
print(POs)
print(IntrinsicVs)
print(IntrinsicVs-POs)
data={"Strike":K,"Maturity":T,"Price":POs,"K-S":IntrinsicVs,"Intrinsic-Price":IntrinsicVs-POs}
df=pd.DataFrame(data,columns=["Strike","Maturity","Price","K-S","Intrinsic-Price"])
print(df)

8) Show that the premium of the price of a European call option over its intrinsic value max(S - K, 0) is largest at the money. In other words, show that the maximum value of  
CBS(S) - max(S - K,O)  
is obtained for S = K, where CBS(S) is the Black-Scholes value of the plain vanilla European call option with strike K and spot price S.  

$Prem=Se^{-qT}N(d_1)-Ke^{-rT}N(d_2)-max(S-K,0)$  
Delta typically is $N(d_1)$ if $q==0$.  
Probability of S<K is $N(d_2)$  
$P=Se^{-qT}N(d_1)-Ke^{-rT}N(d_2)-S+K\qquad$ if $S>K$  
$\qquad=Se^{-qT}N(d_1)-Ke^{-rT}N(d_2)\qquad$ if $S\leq K$  
$P'=e^{-qT}N(d_1)-1\qquad$ if $S>K$  
$\qquad   <N(d_1)-1$  
$\qquad   <-N(-d_1)$  

Thus P is a decreasing function when S>K. And since the other leg is a pure Call Option price it is an increasing function when S<K. Thus the premium over the intrinsic pay off is maximum at S=K.

Thus an American Option Call Option is same price as an OTM Call Option till it is ATM. --> Verify this statement.


9) Use formula  $V(S,K,t) = C(S,K,t) - (\frac{B}{S})^{2a}C(\frac{B^2}{S},K,t)$ to price a six months down-and-out call on a non- dividend-paying asset with price following a lognormal distribution with 30% volatility and spot price 40. The barrier is B = 35 and the strike for the call is K = 40. The risk-free interest rate is constant at 5%.

In [ ]:
import math
from scipy.stats import norm
S=40
B=35
K=40
r=0.05
q=0
v=0.3
T=0.5
d1=lambda S,K,r,q,v,T:(math.log(S/K)+(r-q+v**2/2)*T)/v/T**0.5
d2=lambda S,K,r,q,v,T:(math.log(S/K)+(r-q-v**2/2)*T)/v/T**0.5
C=lambda S,K,r,q,v,T:S*math.exp(-q*T)*norm.cdf(d1(S,K,r,q,v,T))-K*math.exp(-r*T)*norm.cdf(d2(S,K,r,q,v,T))

CS=C(S,K,r,q,v,T)
CB=C(B**2/S,K,r,q,v,T)
a=(r-q)/v**2-1/2

print("Value of the down and out call option is:{:.2f}".format(CS-(B/S)**(2*a)*CB))
print("Value of the regular call option is:{:.2f}".format(CS))

    


10) Show that the value of a down-and-out call with barrier B less than the strike K of the call, i.e., B < K, converges to the value of a plain vanilla call with strike K when $B \rightarrow 0$. For simplicity, assume that the underlying asset does not pay dividends and that interest rates are zero.

$V(S,K,B,r,q,v,T)=C(S,K,r,q,v,T)-(\frac{B}{S})^{2a}C(\frac{B^2}{S},K,r,q,v,T)$  
$q=r=0 \implies$  
$V(S,K,B,v,T)=C(S,K,v,T)-(\frac{B}{S})^{2a}C(\frac{B^2}{S},K,v,T)$  

$C(\frac{B^2}{S},K,v,T)=\frac{B^2}{S}N(d_1')-KN(d_2')$  
Given K,B,v and T
$a=\frac{r-q}{\sigma^2} -\frac{1}{2}=-\frac{1}{2}$  
$V(S)=C(S)-\frac{S}{B}(\frac{B^2}{S}N(d_1')-KN(d_2'))$  
$V(S)=C(S)-(BN(d_1')-\frac{SK}{B}N(d_2'))$  
$V(S)=C(S)-BN(d_1')+\frac{SK}{B}N(d_2')$  

Now we need ot prove that   
$Lt_{B\rightarrow 0}\frac{SK}{B}N(d_2')=0$  
$Lt_{B\rightarrow 0}\frac{SK}{B\sqrt{2\pi}}\int_{-\infty}^{d_2'}e^\frac{-x^2}{2}\,dx =0$  
As $B\rightarrow 0$   $d_2'=\frac{log\left(\frac{B^2}{SK}\right)-\frac{\sigma^2}{2}T}{\sigma\sqrt T} \rightarrow -\infty$ and the integral in above $\rightarrow 0$ as the integral is now from $-\infty$ to $-\infty$  

Applying L'Hospitals the limit becomes  
$Lt_{B\rightarrow 0}\frac{SK}{\sqrt{2\pi}}e^\frac{-{d_2'}^2}{2}\frac{dd_2'}{dB}$    
$Lt_{B\rightarrow 0}\frac{SK}{\sqrt{2\pi}}e^\frac{-{d_2'}^2}{2}\frac{2BSK}{B^2SK\sigma\sqrt{T}}$    
$Lt_{B\rightarrow 0}\frac{2SK}{\sigma B\sqrt{2\pi T}}e^\frac{-{d_2'}^2}{2}$   
As $B\rightarrow 0\,\,\,\,d_2'\rightarrow-\infty\,\,\,\,d_2'^2\rightarrow\infty\,\,\,\,e^\frac{-{d_2'}^2}{2}\rightarrow0$  

Thus a down and out barrier option would have it's price approach a vanilla call option price when $B\rightarrow0$  


11) Compute the Delta and Gamma of a down-and-out call with B < K, using formula

$V(S,K,t) = C(S,K,t) - (\frac{B}{S})^{2a}C(\frac{B^2}{S},K,t)$  
$\frac{dd_1'}{dS}=-\frac{SB^2}{\sigma B^2\sqrt{T}S^2}=-\frac{1}{\sigma S \sqrt{T}}$   
$\frac{V(S,K,t)}{dS} = e^{-qT}N(d_1) - 2a(\frac{B}{S})^{2a-1}\left(-\frac{B}{S^2}\right)C(\frac{B^2}{S},K,t) - (\frac{B}{S})^{2a}e^{-qT}N(d_1')\left(-\frac{B^2}{S^2}\right)$  
$\Delta = e^{-qT}N(d_1) + \frac{2a}{S}(\frac{B}{S})^{2a}C(\frac{B^2}{S},K,t) + \frac{B^2}{S^2}\left(\frac{B}{S}\right)^{2a}e^{-qT}N(d_1')$   
$\Delta = e^{-qT}N(d_1) + \left(\frac{B}{S}\right)^{2a}\left(\frac{2a}{S}C(\frac{B^2}{S},K,t) + \frac{B^2}{S^2}e^{-qT}N(d_1')\right)$  
$\Delta = e^{-qT}N(d_1) + \frac{2aB^{2a}}{S^{2a+1}}C(\frac{B^2}{S},K,t) + \frac{B^{2a+2}}{S^{2a+2}}e^{-qT}N(d_1')$   
$\Gamma=e^{-qT}N'(d1)\frac{dd_1}{dS} -(2a+1)\frac{2aB^{2a}}{S^{2a+2}}C(\frac{B^2}{S},K,t) + \frac{2aB^{2a}}{S^{2a+1}}e^{-qT}N(d_1')\left(-\frac{B^2}{S^2}\right) -(2a+2)\frac{B^{2a+2}}{S^{2a+3}}e^{-qT}N(d_1')+ \frac{B^{2a+2}}{S^{2a+2}}e^{-qT}N'(d_1')\left(-\frac{1}{\sigma S \sqrt{T}}\right)$  
$\Gamma=e^{-qT}N'(d1)\frac{1}{\sigma S\sqrt{T}} -2a(2a+1)\frac{B^{2a}}{S^{2a+2}}C(\frac{B^2}{S},K,t) -(4a+2)\frac{B^{2a+2}}{S^{2a+3}}e^{-qT}N(d_1')+ \frac{B^{2a+2}}{S^{2a+2}}e^{-qT}N'(d_1')\left(-\frac{1}{\sigma S \sqrt{T}}\right)$  
$\Gamma=\frac{1}{\sigma S\sqrt{2\pi T}}e^{-qT-\frac{d_1^2}{2}} -2a(2a+1)\frac{B^{2a}}{S^{2a+2}}C(\frac{B^2}{S},K,t) -(4a+2)\frac{B^{2a+2}}{S^{2a+3}}e^{-qT}N(d_1')- \frac{1}{\sigma S \sqrt{2\pi T}}$ 
$\frac{B^{2a+2}}{S^{2a+2}}e^{-qT-\frac{d_{1p}^2}{2}}$  




### Chapter 8

In [ ]:
P=lambda n,T:np.prod(np.array([1] if n==T else [math.pi*i**2/(math.pi*i**2-1)for i in range(n+1,T+1)]))
S=lambda n,T:1/(1-math.pi*n**2)*P(n,T)

for i in range(20):
    print(S(i,20))
print(sum([S(i,20) for i in range(21)]))


** Bisection Method **  
Example: Use the bisection method on the interval [-2,3] to find a zero of the function:  
$f(x) = x^4 -5x^2 +4-\frac{1}{1+e^{x^3}}$  

Ans:
First make sure that $f^{'}(x)$ exists  
$f^{'}(x)=4x^3-10x+\frac{3x^2e^{x^3}}{(1+e^{x^3})^2}$   
Thus $f^{'}(x)$ exists in [-2,3]
make sure f(-2) and f(-3) has different signs


In [ ]:
f=lambda x:x**4-5*x**2+4-1/(1+math.exp(x**3))
print(f(-2))
print(f(3))

int_tol=10**(-6)
tol_app=10**(-9)
l=-2
r=3
m=0
f_curr=0
tol=100
interval=r-l
while interval>int_tol and tol>tol_app:
    interval=r-l
    m=(l+r)/2
    f_curr=f(m)
    if f_curr<0 and f(l)<0:
        l=m
    else:
        r=m
    interval = r-l
    tol=abs(f_curr)
print(m)
print(f_curr)
    
    

** Newton's method ** 


In [ ]:
f=lambda x:x**4-5*x**2+4-1/(1+math.exp(x**3))
f1=lambda x:4*x**3-10*x+3*x**2*math.exp(-x**3)/(1+math.exp(-x**3))**2


int_tol=10**(-6)
tol_app=10**(-9)
interval=3
#Any number below this will make the x^3>709.7823 for which the current python fails to calc exp(x^3)
oldx=newx=0.03946502
tol=abs(f(oldx))
while interval>int_tol and tol>tol_app:
    print("f(x):f1(x)={:.4f}:{:.4f}".format(f(oldx),f1(oldx)))
    newx=oldx-f(oldx)/f1(oldx)
    print("pow(x,3)={:.4f}".format(newx**3))
    print("x:f(x)={:.4f}:{:.4f}".format(newx,f(newx)))
    interval=abs(newx-oldx)
    tol=abs(f(newx))
    oldx=newx
print(newx)
print(f(newx))

**Secant Method**  


In [ ]:
f=lambda x:x**4-5*x**2+4-1/(1+math.exp(x**3))
f1=lambda x1,x2,fx1,fx2:(fx2-fx1)/(x2-x1)


int_tol=10**(-6)
tol_app=10**(-9)
interval=3
#Any number below this will make the x^3>709.7823 for which the current python fails to calc exp(x^3)
oox=ox=nx=0.03946502
oox=ox=nx=-0.5
fox=f(ox)
oox=ox+0.01
foox=f(oox)
tol=abs(f(ox))
while interval>int_tol and tol>tol_app:
    f1ox=f1(ox,oox,fox,foox)
    print("f(x):f1(x)={:.4f}:{:.4f}".format(fox,f1ox))
    nx=ox-fox/f1ox
    fnx=f(nx)
    print("x:f(x)={:.4f}:{:.4f}".format(nx,fnx))
    interval=abs(nx-ox)
    tol=abs(f(nx))
    oox=ox
    foox=fox
    ox=nx
    fox=fnx

print(nx)
print(fnx)

Use Newton's method to solve F(x) = 0, for  
$F(x)=\left[\begin{array}{c} x_1^3+2x_1x_2+x_3^2-x_2x_3+9\\
                             2x_1^2+2x_1x_2^2+x_2^3x_3^2-x_2^2x_3-2\\
                             x_1x_2x_3+x_1^3-x_3^2-x_1x_2^2-4\end{array}\right]$  

$\nabla F=\left[\begin{array}{ccc} 
3x_1^2+2x_2                           & 2x_1-x_3                           & 2x_3-x_2\\
4x_1+2x_2^2                           & 4x_1x_2+3x_2^2x_3^2-2x_2x_3        & 2x_2^3x_3-x_2^2\\
x_2x_3+3x_1^2-x_2^2                   & x_1x_3-2x_1x_2                     &x_1x_2-2x_3
\end{array}\right]$  


In [ ]:
f=lambda x:np.array([x[0]**3       +2*x[0]*x[1]   +x[2]**2        -x[1]*x[2]   +9,
            2*x[0]**2     +2*x[0]*x[1]**2+x[1]**3*x[2]**2-x[1]**2*x[2]-2,
            x[0]*x[1]*x[2]+x[0]**3       -x[2]**2        -x[0]*x[1]**2-4])
gf=lambda x:np.array([
[3*x[0]**2 + 2*x[1]              , 2*x[0]       - x[2]                            , 2*x[2]         - x[1]    ],
[4*x[0]    + 2*x[1]**2           , 4*x[0]*x[1] + 3*x[1]**2*x[2]**2 - 2*x[1]*x[2] , 2*x[1]**3*x[2] - x[1]**2 ],
[x[1]*x[2] + 3*x[0]**2 - x[1]**2 , x[0]*x[2]   - 2*x[0]*x[1]                     , x[0]*x[1]      - 2*x[2]  ]
])

norm=lambda x:sum(x*x)**0.5

nx=ox=np.array([1,2,3])
fnx=fox=f(ox)
tol=norm(fox)
interval=norm(ox)
tol_app=10**(-9)
int_tol=10**(-6)
iter=0
while tol>tol_app and interval>int_tol:
    iter=iter+1
    nx= ox-np.linalg.solve(gf(ox),fox)
    print("at iter {} new x is {}".format(iter,nx))
    fnx=f(nx)
    interval=norm(nx-ox)
    tol=norm(fnx)
    ox=nx
    fox=fnx
print(nx)
print(fnx)
print("Number of iterations is {}".format(iter))

Same solution with finite difference method

In [ ]:
f=lambda x:np.array([x[0]**3       +2*x[0]*x[1]   +x[2]**2        -x[1]*x[2]   +9,
            2*x[0]**2     +2*x[0]*x[1]**2+x[1]**3*x[2]**2-x[1]**2*x[2]-2,
            x[0]*x[1]*x[2]+x[0]**3       -x[2]**2        -x[0]*x[1]**2-4])
gf=lambda x,fx:np.c_[(f(x+np.array([int_tol,0,0]))-fx)/int_tol,(f(x+np.array([0,int_tol,0]))-fx)/int_tol,(f(x+np.array([0,0,int_tol]))-fx)/int_tol]

norm=lambda x:sum(x*x)**0.5

nx=ox=np.array([2,2,2])
fnx=fox=f(ox)
tol=norm(fox)
interval=norm(ox)
tol_app=10**(-9)
int_tol=10**(-6)
iter=0
while tol>tol_app and interval>int_tol:
    iter=iter+1
    #print((f(ox+np.array([int_tol,0,0]))-fox)/int_tol)
    #print(gf(ox,fox))
    nx= ox-np.linalg.solve(gf(ox,fox),fox)
    #print("at iter {} new x is {}".format(iter,nx))
    fnx=f(nx)
    interval=norm(nx-ox)
    tol=norm(fnx)
    ox=nx
    fox=fnx
print(nx)
print(fnx)
print("Number of iterations is {}".format(iter))

Find a minimal variance portfolio with 11:5% .expected rate. of return, if four assets can be traded to set up the portfoho, gIVen the follOWIng data on the rates of return of the assets:
The solution to this system is  
$\mu_1=0.09;\,\,\sigma_1=0.2;\,\,\rho_{1,2}=-0.5$  
$\mu_2=0.12;\,\,\sigma_2=0.3;\,\,\rho_{2,3}=0.25$  
$\mu_3=0.15;\,\,\sigma_1=0.35;\,\,\rho_{1,3}=0.35$  
$\mu_4=0.06;\,\,\sigma_1=0.15;\,\,\rho_{i,4}=0$  

$\mu_p=0.115$  
Problem is to reduce variance
$v(w_1,w_2,w_3,w_4)=w_1^20.2^2+w_2^20.3^2+w_3^20.35^2+w_4^20.15^2 + 2w_1w_2*0.2*0.3*-0.5 + 2w_1w_3*0.2*0.35*0.35 +2w_2w_3*0.3*0.35*0.25$  
Given  
$w_1+w_2+w_3+w_4=1$  
$0.09w_1+0.12w_2+0.15w_3+0.06w_4=0.115$  
Set of Equations are  
$g_1=w_1+w_2+w_3+w_4-1$  
$g_2=0.09w_1+0.12w_2+0.15w_3+0.06w_4-0.115$  
$v(w_1,w_2,w_3,w_4)=0.04w_1^2+0.09w_2^2+0.1225w_3^2+0.0225w_4^2 - 0.06w_1w_2 + 0.049w_1w_3 +0.0525w_2w_3$  

$F(w_1,w_2,w_3,w_4,\lambda_1,\lambda_2)=v+\lambda_1g_1+\lambda_2g_2$  

$\nabla F=\left[\begin{array}{c} 
0.08w_1-0.06w_2+0.049w_3+\lambda_1+0.09\lambda_2\\
0.18w_2-0.06w_1+0.0525w_3+\lambda_1+0.12\lambda_2\\
0.245w_3+0.049w_1+0.0525w_2+\lambda_1+0.15\lambda_2\\
0.045w_4+\lambda_1+0.06\lambda_2\\
w_1+w_2+w_3+w_4-1\\
0.09w_1+0.12w_2+0.15w_3+0.06w_4-0.115
\end{array}\right]=0$  

$w_1=1-w_2-w_3-w_4$  
$0.09w_1+0.12w_2+0.15w_3+0.06w_4-0.115=0\\
\implies0.09(1-w_2-w_3-w_4)+0.12w_2+0.15w_3+0.06w_4-0.115=0\\
\implies 0.03w_2+0.06w_3-0.03w_4-0.025=0\\
\implies w_2=5/6-2w_3+w_4\\
\implies w_1=1-(5/6-2w_3+w_4)-w_3-w_4\\
\implies w_1=1/6+w_3-2w_4$  

$0.08w_1-0.06w_2+0.049w_3+\lambda_1+0.09\lambda_2=0\\
\implies0.08(1/6+w_3-2w_4)-0.06(5/6-2w_3+w_4)+0.049w_3+\lambda_1+0.09\lambda_2=0\\  
\implies0.08(1/2+3w_3-6w_4)-0.06(5/2-6w_3+3w_4)+0.147w_3+3\lambda_1+0.27\lambda_2=0\\
\implies0.04+0.24w_3-0.48w_4-0.15+0.36w_3-0.18w_4+0.147w_3+3\lambda_1+0.27\lambda_2=0\\
\implies0.04-0.15+0.24w_3+0.36w_3+0.147w_3-0.48w_4-0.18w_4+3\lambda_1+0.27\lambda_2=0\\
\implies-0.11+0.747w_3-0.66w_4+3\lambda_1+0.27\lambda_2=0$  

$-0.06w_1+0.18w_2+0.0525w_3+\lambda_1+0.12\lambda_2=0\\
\implies-0.06(1/6+w_3-2w_4)+0.18(5/6-2w_3+w_4)+0.0525w_3+\lambda_1+0.12\lambda_2=0\\
\implies-0.01-0.06w_3+0.12w_4+0.15-0.36w_3+0.18w_4+0.0525w_3+\lambda_1+0.12\lambda_2=0\\
\implies-0.01+0.15-0.06w_3+0.0525w_3-0.36w_3+0.12w_4+0.18w_4+\lambda_1+0.12\lambda_2=0\\
\implies0.14-0.3675w_3+0.30w_4+\lambda_1+0.12\lambda_2=0$  


$0.245w_3+0.049w_1+0.0525w_2+\lambda_1+0.15\lambda_2=0\\
\implies +0.049w_1+0.0525w_2+0.245w_3+\lambda_1+0.15\lambda_2=0\\
\implies +0.049(1/6+w_3-2w_4)+0.0525(5/6-2w_3+w_4)+0.245w_3+\lambda_1+0.15\lambda_2=0\\
\implies +0.049(1+6w_3-12w_4)+0.0525(5-12w_3+6w_4)+1.470w_3+6\lambda_1+0.9\lambda_2=0\\
\implies 0.049+ 0.294w_3-0.588w_4+0.2625-0.63w_3+0.3150w_4+1.470w_3+6\lambda_1+0.9\lambda_2=0\\
\implies 0.049+0.2625 +0.294w_3-0.63w_3+1.470w_3-0.588w_4+0.3150w_4+6\lambda_1+0.9\lambda_2=0\\
\implies 0.3115 +1.134w_3-0.273w_4+6\lambda_1+0.9\lambda_2=0$  

$0.045w_4+\lambda_1+0.06\lambda_2=0$

Now this is a linear system of equations in $w_3,w_4,\lambda_1$ and $\lambda_2$  

Solve for

$\left[\begin{array}{cccc}
0.747&-0.66&+3&+0.27\\
-0.3675&+0.30&+1&+0.12\\
+1.134&-0.273&+6&+0.9\\
0&0.045&+1&+0.06
\end{array}\right]\left[\begin{array}{c}w_3\\w_4\\\lambda_1\\\lambda_2\end{array}\right]=\left[\begin{array}{c}0.11\\-0.14\\-0.3115\\0\end{array}\right]$

$w_3=0.13504227$  
$w_4=-0.12287165$  
$\lambda_1=0.06456892$  
$\lambda_2=-0.9839949$  
$w_1=1/6+w_3-2w_4$
$w_1=0.5474522427456$  
$w_2=0.44037713708344806$

Now let $\nabla g(w,\lambda)v=0$ to get constrained v vector.  

$g=\left[\begin{array}{c}w_1+w_2+w_3+w_4-1\\
0.09w_1+0.12w_2+0.15w_3+0.06w_4-0.115
\end{array}\right]$  
$\nabla g=\left[\begin{array}{cccc}1&1&1&1\\
0.09&0.12&0.15&0.06\end{array}\right]$  

Let's get Hessian of F  
\

$H=\left[\begin{array}{cccc}0.08&-0.06&0.049&0\\
-0.06&0.18&0.0525&0\\
0.049&0.0525&0.245&0\\
0&0&0&0.045\end{array}\right]$  

In [ ]:
import numpy as np
A=np.array([[0.747,-0.66,+3,+0.27],
[-0.3675,+0.30,+1,+0.12],
[+1.134,-0.273,+6,+0.9],
[0,0.045,+1,+0.06]])
b=np.array([0.11,-0.14,-0.3115,0])
ans=np.linalg.solve(A,b)
print(ans)
print("w1={}".format(1/6.+ans[0]-2*ans[1]))
print("w2={}".format(5/6-2*ans[0]+ans[1]))

Hess=np.array([[0.08,-0.06,0.049,0],[-0.06,0.18,0.0525,0],[0.049,0.0525,0.245,0],[0,0,0,0.045]])
print(np.linalg.det(Hess))

A 34 months semiannual coupon bond with coupon rate 8% is priced at 105. The face value of the bond is assumed to be 100. Compute the yield of the bond

In [ ]:
import numpy as np
import math

T=34/12.
C=0.08
B=105
Freq=2
F=100
ti=np.arange(T,0,-1/Freq)[::-1]
print(ti)
cf=np.full((len(ti)),C*F)
cf[-1]+=F
print(cf)

f=lambda y:sum([cf[i]*math.exp(-y*ti[i]) for i in range(len(cf))])-B
f1=lambda y:sum([cf[i]*ti[i]*math.exp(-y*ti[i]) for i in range(len(cf))])

tol_app=10**(-9)
int_tol=10**(-6)


oy=0.1
interval=oy
foy=f(oy)

while foy>tol_app or interval>int_tol:
    ny=oy+foy/f1(oy)
    print("y:P={:.6f}:{:.6f}".format(ny,foy+B))
    interval=ny-oy
    oy=ny
    foy=f(oy)

print(oy)
print(foy)




Example: The price of a one year call option with strike 20 on a non-dividend paying asset with spot price 25 is \$7. Assuming that the risk-free interest rate is constant at 5\% over the life of the option, find the implied volatility  
$C=SN(d_1)-Ke^{-rT}N(d_2)$  
$\kappa=SN^{'}(d_1)\left(\frac{dd_1}{d\sigma}-\frac{dd_2}{d\sigma}\right)$  
$\frac{dd_1}{d\sigma}=\frac{\sqrt{T}}{2}$  
$\left(\frac{dd_1}{d\sigma}-\frac{dd_2}{d\sigma}\right)=\sqrt{T}$  
$\kappa=S\sqrt{\frac{T}{2\pi}}e^{-\frac{d_1^2}{2}}$  

In [3]:
from scipy.stats import norm
import math
K=20
S=25
C=7
r=0.05
v=0.25
T=1
q=0

d1=lambda S,K,r,q,v,T:(math.log(S/K)+(r+q+v**2/2)*T)/v/T**0.5
d2=lambda S,K,r,q,v,T:(math.log(S/K)+(r+q-v**2/2)*T)/v/T**0.5
f=lambda v:ff(S,K,r,q,v,T)
ff=lambda S,K,r,q,v,T:S*norm.cdf(d1(S,K,r,q,v,T))-K*math.exp(-r*T)*norm.cdf(d2(S,K,r,q,v,T))-C
f1=lambda v:ff1(S,K,r,q,v,T)
ff1=lambda S,K,r,q,v,T:S*(T/2/math.pi)**0.5*math.exp(-d1(S,K,r,q,v,T)**2/2)

tol_app=10**(-9)
int_tol=10**(-6)
ov=0.25
fov=f(ov)
interval=ov

while abs(fov)>tol_app or interval>int_tol:
    nv=ov-fov/f1(ov)
    print("v:C={:.7f}:{:.7f}".format(ov,fov+C))
    interval=abs(nv-ov)
    ov=nv
    fov=f(ov)
print(ov)
print(fov+C)


v:C=0.2500000:6.3531280
v:C=0.3861088:7.1509919
v:C=0.3634007:7.0021771
v:C=0.3630633:7.0000005
0.3630631804856223
7.000000000000034


Exercise 1  
Find the maximum and minimum ofthe function $f(x_1,x_2, x_3) = 4x_2- 2x_3$ subject to the constraints $2x_1 - x_2 - x_3 = 0$ and $x_1^2+x_2^2=13$.

$F=4x_2- 2x_3+\lambda_1(2x_1 - x_2 - x_3)+\lambda_2(x_1^2+x_2^2-13)$  
$\nabla F=\left[\begin{array}{c}
2\lambda_1+2\lambda_2x_1\\
4-\lambda_1+2\lambda_2x_2\\
-2-\lambda_1\\
2x_1-x_2-x_3\\
x_1^2+x_2^2-13\end{array}\right]$  



$\lambda_1=-2$  
$\lambda_2x_1=2$  
$\lambda_2x_2=-3$  
$x_1=\frac{2}{-3}x_2$  
$\frac{4}{9}x_2^2+x_2^2=13$  
$4x_2^2+9x_2^2=13*9$ 

$x_2=3$  
$x_1=-2$  
$\lambda_2=-1$  
$-7-x_3=0$  
$x_3=-7$  

$x_2=-3$
$x_1=+2$  
$\lambda_2=1$  
$x_3=7$  



F(-2,3,-7)=12-14=-2
F(2,-3,7)=-12-14=-26

$H(v,\lambda_1,\lambda_2)=\left[\begin{array}{ccc}
2\lambda_2&0&0\\
0&2\lambda_2&0\\
0&0&0\end{array}\right]$  

$H(v,-2,-1)=\left[\begin{array}{ccc}
-2&0&0\\
0&-2&0\\
0&0&0\end{array}\right]$  
$\implies -2x_1^2-2x_2^2\leq0$  
Thus F(-2,3,-7)=12-14=-2 is the maximum extrema.


Assume that you can trade four assets (and that it is also possible to short the assets). The expected values, standard deviations, and correlations of the rates of return of the assets are:

$\mu_1=0.08;\,\,\sigma_1=0.25;\,\,\rho_{1,2}=-0.25$  
$\mu_2=0.12;\,\,\sigma_2=0.25;\,\,\rho_{2,3}=-0.25$  
$\mu_3=0.16;\,\,\sigma_3=0.30;\,\,\rho_{1,3}=0.25$  
$\mu_4=0.06;\,\,\sigma_4=0.15;\,\,\rho_{i,4}=0$  

(i) Find the asset allocation for a minimal variance portfolio with 12% expected rate of return;  
(ii) Find the asset allocation for a maximum expected return portfolio with standard deviation of the rate of return equal to 24%  

$w_1,w_2,w_3,w_4$ be the allocation for the portfolio.

The constraings g(w) are  
$w_1+w_2+w_3+w+4=1$  
$0.08w_1+0.12w_2+0.16w_3+0.06w_4=0.12$  

Minimize
$v=0.25^2w_1^2+0.25^2w_2^2+0.30^2w_3^2+0.15^2w_4^2+0.25*0.25*-0.25w_1w_2+0.25*0.30*0.25w_1w_3+0.25*0.30*-0.25w_2w_3$  
$v=0.0625w_1^2+0.0625w_2^2+0.09w_3^2+0.0225w_4^2-0.015625w_1w_2+0.01875w_1w_3-0.01875w_2w_3$  
$g_1=w_1+w_2+w_3+w_4-1$  
$g_2=0.08w_1+0.12w_2+0.16w_3+0.06w_4-0.12$  

Lagrange  
$F=v+\lambda_1g_1+\lambda_2g_2$  
$\nabla g=\left[\begin{array}{cccc} 
1&1&1&1\\
0.08&0.12&0.16&0.06
\end{array}\right]$  
This shows that g has full rank in m and therefore can give m distinct solutions enabling a solution for Lagrange.

Gradient of F is  
$\nabla F=\left[\begin{array}{c} 
2*0.0625w_1-0.015625w_2+0.01875w_3+\lambda_1+0.08\lambda_2\\
2*0.0625w_2-0.015625w_1-0.01875w_3+\lambda_1+0.12\lambda_2\\
2*0.09w_3+0.01875w_1-0.01875w_2+\lambda_1+0.16\lambda_2\\
2*0.0225w_4++\lambda_1+0.06\lambda_2\\
w_1+w_2+w_3+w_4-1\\
0.08w_1+0.12w_2+0.16w_3+0.06w_4-0.12
\end{array}\right]$

$\nabla F=\left[\begin{array}{c} 
0.125w_1-0.015625w_2+0.01875w_3+\lambda_1+0.08\lambda_2\\
0.125w_2-0.015625w_1-0.01875w_3+\lambda_1+0.12\lambda_2\\
0.18w_3+0.01875w_1-0.01875w_2+\lambda_1+0.16\lambda_2\\
0.045w_4++\lambda_1+0.06\lambda_2\\
w_1+w_2+w_3+w_4-1\\
0.08w_1+0.12w_2+0.16w_3+0.06w_4-0.12
\end{array}\right]=0\\
\implies \left[\begin{array}{c} 
0.125w_1    -0.015625w_2  +0.01875w_3+   0    +\lambda_1+0.08\lambda_2\\
-0.015625w_1+0.125w_2     -0.01875w_3+   0    +\lambda_1+0.12\lambda_2\\
+0.01875w_1 -0.01875w_2   +0.18w_3   +   0    +\lambda_1+0.16\lambda_2\\
      0     +    0        +   0      +0.045w_4+\lambda_1+0.06\lambda_2\\
     w_1    +   w_2       + w_3      +w_4                                -1\\
  0.08w_1   +  0.12w_2    +0.16w_3   +0.06w_4                            -0.12
\end{array}\right]=0\\
\implies \left[\begin{array}{cccccc} 
0.125       & -0.015625   &0.01875   &   0    &   1      &0.08\\
-0.015625   &   0.125     &-0.01875  &   0    &   1      &0.12\\
  0.01875   &  -0.01875   &   0.18   &   0    &   1      &0.16\\
      0     &    0        &   0      &0.045   &   1      &0.06\\
     1      &    1        &   1      &   1    &   0      &  0 \\
  0.08      &   0.12      & 0.16     &  0.06  &   0      &  0
\end{array}\right]\left[\begin{array}{c}w_1\\w_2\\w_3\\w_4\\\lambda_1\\\lambda_2\end{array}\right]=\left[\begin{array}{c}0\\0\\0\\0\\1\\0.12\end{array}\right]$

In [4]:
import numpy as np
a=np.array([[0.125    ,-0.015625,0.01875 ,0    ,1,0.08],
            [-0.015625,0.125    ,-0.01875,0    ,1,0.12],
            [0.01875  ,-0.01875 ,0.18    ,0    ,1,0.16],
            [0        ,0        ,0       ,0.045,1,0.06],
            [1        ,1        ,1       ,1    ,0,0   ],
            [0.08     ,0.12     ,0.16    ,0.06 ,0,0   ]])
b=[0,0,0,0,1,0.12]
print(np.linalg.solve(a,b))

[ 0.13006586  0.37164192  0.35100167  0.14729054  0.02458553 -0.52022676]


$\implies \left[\begin{array}{c}w_1\\w_2\\w_3\\w_4\\\lambda_1\\\lambda_2\end{array}\right]=\left[\begin{array}{c}0.13006586\\  0.37164192\\  0.35100167\\  0.14729054\\  0.02458553\\ -0.52022676\end{array}\right]$  



Now to decide if the extremum point reaches minima or maxima, we need to use Hessian.  
First $\nabla g v=0$  
$\left[\begin{array}{cccc} 
1&1&1&1\\
0.08&0.12&0.16&0.06
\end{array}\right]\left[\begin{array}{c}w_1\\w_2\\w_3\\w_4\end{array}\right]=0$  
$w_1=-w_2-w_3-w_4$  
$0.08w_1+0.12w_2+0.16w_3+0.06w_4=0\\
\implies(-0.08w_2-0.08w_3-0.08w_4)+0.12w_2+0.16w_3+0.06w_4=0\\  
\implies(-0.08w_2-0.08w_3-0.08w_4)+0.12w_2+0.16w_3+0.06w_4=0\\
\implies+0.12w_2-0.08w_2+0.16w_3-0.08w_3+0.06w_4-0.08w_4=0\\
\implies+0.04w_2+0.08w_3-0.02w_4=0\\
\implies w_4=2w_2+4w_3$  
$w_1=-w_2-w_3-w_4\\
\implies w_1=-w_2-w_3-(2w_2+4w_3)\\
\implies w_1=-3w_2-5w_3$  

$H(v,\lambda_1,\lambda_2)=\nabla \left[\begin{array}{c} 
0.125w_1-0.015625w_2+0.01875w_3+\lambda_1+0.08\lambda_2\\
0.125w_2-0.015625w_1-0.01875w_3+\lambda_1+0.12\lambda_2\\
0.18w_3+0.01875w_1-0.01875w_2+\lambda_1+0.16\lambda_2\\
0.045w_4++\lambda_1+0.06\lambda_2
\end{array}\right]\\
=\left[\begin{array}{cccc} 
0.125     & -0.015625  & +0.01875 & 0\\
-0.015625 & 0.125      &-0.01875  & 0\\
0.01875   &-0.01875    &+0.18     & 0\\
0         & 0          &   0      &0.045
\end{array}\right]$

$v^THv=v_1(0.125v_1−0.015625v_2+0.01875v_3) +v_2(−0.015625v_1+0.125v_2−0.01875v_3)+v_3(0.01875v_1−0.01875v_2+0.18v_3)+0.045v_4^2\\ \quad=0.125v_1^2−0.015625v_1v_2+0.01875v_1v_3−0.015625v_1v_2+0.125v_2^2−0.01875v_2v_3+0.01875v_1v_3−0.01875v_2v_3+0.18v_3^2+0.045v_4^2\\
\quad=0.125v_1^2−0.015625v_1v_2−0.015625v_1v_2+0.01875v_1v_3+0.01875v_1v_3 +0.125v_2^2−0.01875v_2v_3−0.01875v_2v_3+0.18v_3^2+0.045v_4^2\\
\quad=0.125v_1^2−0.03125v_1v_2+0.03750v_1v_3 +0.125v_2^2−0.0375v_2v_3+0.18v_3^2+0.045v_4^2\\
\quad=0.125v_1^2+0.00125v_1(−25v_2+21v_3) +0.125v_2^2−0.0375v_2v_3+0.18v_3^2+0.045v_4^2\qquad mul\,by\,10^5\\
\quad=12500v_1^2+125v_1(−25v_2+21v_3) +12500v_2^2−3750v_2v_3+18000v_3^2+4500v_4^2\qquad div\,by\,25\\ 
\quad=500v_1^2+5v_1(−25v_2+21v_3) +500v_2^2−150v_2v_3+720v_3^2+180v_4^2\qquad div\,by\,5\\
\quad=100v_1^2−25v_1v_2+21v_1v_3+100v_2^2−30v_2v_3+144v_3^2+36v_4^2\qquad mul\,by\,2\\\\
\quad=200v_1^2−2*25v_1v_2+2*21v_1v_3+200v_2^2−60v_2v_3+288v_3^2+72v_4^2\\
\quad=25v_1^2−2*25v_1v_2+25v_2^2   +9v_1^2+2*21v_1v_3+49v_3^2   +25v_2^2−2*5*15v_2v_3+225v_3^2      +166v_1^2+150v_2^2+14v_3^2+72v_4^2\\
\quad=25(v_1-v_2)^2   +(3v_1+7v_3)^2   +25(v_2-3v_3)^2      +166v_1^2+150v_2^2+14v_3^2+72v_4^2$  
For all $v_1\neq v_2\neq v_3 \neq v4\neq 0$ the hessian is greater than zero.
Thus the variance is minimum at this solution.  

Answer:$w_1=0.13006586\,\,w_2=  0.37164192\,\,  w_3=0.35100167\,\,w_4=  0.14729054$  

The constraings g(w) are  
$w_1+w_2+w_3+w_4=1$  
$0.25^2w_1^2+0.25^2w_2^2+0.30^2w_3^2+0.15^2w_4^2+0.25*0.25*-0.25w_1w_2+0.25*0.30*0.25w_1w_3+0.25*0.30*-0.25w_2w_3=0.24^2$

Maximize  
$v=0.08w_1+0.12w_2+0.16w_3+0.06w_4$  
$g_1=w_1+w_2+w_3+w_4-1$  
$g_2=0.0625w_1^2+0.0625w_2^2+0.09w_3^2+0.0225w_4^2-0.015625w_1w_2+0.01875w_1w_3-0.01875w_2w_3-0.0576$  


Lagrange  
$F=v+\lambda_1g_1+\lambda_2g_2$  
$\nabla g=\left[\begin{array}{cccc} 
1&1&1&1\\
0.125w_1-0.015625w_2+0.01875w_3&0.125w_2-0.015625w_1-0.01875w_3& 0.18w_3+0.01875w_1-0.01875w_2&0.045w_4
\end{array}\right]$  
This shows that g has full rank in m and therefore can give m distinct solutions enabling a solution for Lagrange.



$\nabla F=\left[\begin{array}{c}
0.08+\lambda_1+\lambda_2(0.125w_1-0.015625w_2+0.01875w_3)\\
0.12+\lambda_1+\lambda_2(0.125w_2-0.015625w_1-0.01875w_3)\\
0.16+\lambda_1+\lambda_2(0.18w_3+0.01875w_1-0.01875w_2)\\
0.06+\lambda_1+\lambda_2(0.045w_4)\\
w_1+w_2+w_3+w_4-1\\
0.0625w_1^2+0.0625w_2^2+0.09w_3^2+0.0225w_4^2-0.015625w_1w_2+0.01875w_1w_3-0.01875w_2w_3-0.0576
\end{array}\right]=0\\
\implies\left[\begin{array}{c}
0.08+\lambda_1+\lambda_2(0.125w_1-0.015625w_2+0.01875w_3)\\
0.12+\lambda_1+\lambda_2(-0.015625w_1+0.125w_2-0.01875w_3)\\
0.16+\lambda_1+\lambda_2(0.01875w_1-0.01875w_2+0.18w_3)\\
0.06+\lambda_1+\lambda_2(0.045w_4)\\
w_1+w_2+w_3+w_4-1\\
0.0625w_1^2+0.0625w_2^2+0.09w_3^2+0.0225w_4^2-0.015625w_1w_2+0.01875w_1w_3-0.01875w_2w_3-0.0576
\end{array}\right]=0$  

For newton's method we find $\nabla \nabla F$ to get new $nx=ox-(G')^{-1}G$  

$G'=\left[\begin{array}{cccccc}
0.125\lambda_2& -0.015625\lambda_2&0.01875\lambda_2&0&1&0.125w_1-0.015625w_2+0.01875w_3\\
-0.015625\lambda_2&0.125\lambda_2&-0.01875\lambda_2&0&1&-0.015625w_1+0.125w_2-0.01875w_3\\
0.01875\lambda_2&-0.01875\lambda_2&0.18\lambda_2&0&1&0.01875w_1-0.01875w_2+0.18w_3\\
0&0&0&0.045w_4\lambda_2&1&0.045w_4\\
1&1&1&1&0&0\\
0.125w_1-0.015625w_2+0.01875w_3&-0.015625w_1+0.125w_2-0.01875w_3&0.01875w_1-0.01875w_2+0.18w_3&0.045w_4&0&0
\end{array}\right]$  

In [17]:
import numpy as np

int_tol=10**(-6)
tol_app=10**(-9)

f=lambda w1,w2,w3,w4,l1,l2:np.array([0.08+l1+l2*(0.125*w1    -0.015625*w2+0.01875*w3),
                            0.12+l1+l2*(-0.015625*w1+0.125*w2   -0.01875*w3),
                            0.16+l1+l2*(0.01875*w1  -0.01875*w2 +0.18*w3),
                            0.06+l1+l2*(0.045*w4),
                            w1+w2+w3+w4-1,
                            0.0625*w1**2+0.0625*w2**2+0.09*w3**2+0.0225*w4**2-0.015625*w1*w2+0.01875*w1*w3-0.01875*w2*w3-0.0576])
f1=lambda w1,w2,w3,w4,l1,l2:np.array([
                             [0.125*l2        ,-0.015625*l2   ,0.01875*l2    ,0          ,1,0.125*w1    -0.015625*w2+0.01875*w3 ],
                             [-0.015625*l2    ,0.125*l2       ,-0.01875*l2   ,0          ,1,-0.015625*w1+0.125*w2   -0.01875*w3 ],
                             [0.01875*l2      ,-0.01875*l2    ,0.18*l2       ,0          ,1,0.01875*w1  -0.01875*w2 +0.18*w3    ],
                             [0               ,0              ,0             ,0.045*l2   ,1,0.045*w4                            ],
                             [1               ,1              ,1             ,1          ,0,0                                   ],
                             [0.125*w1-0.015625*w2+0.01875*w3,-0.015625*w1+0.125*w2-0.01875*w3,0.01875*w1-0.01875*w2+0.18*w3,0.045*w4,0,0]])
norm=lambda a:sum(a*a)**0.5


ox=np.array([0.0107,0.6450,0.6946,-0.3503,-0.0738,-0.8510])
ox=np.array([0.25,0.25,0.25,0.25,1,1])
fox=f(*ox)
err=norm(fox)
interval=norm(ox)
iter=0
print(sum(ox[:-2]))
print((fox[-1]+0.0576)**0.5)
while interval>int_tol or err>tol_app:
    nx=ox-np.linalg.solve(f1(*ox),fox)
    interval=norm(nx-ox)
    ox=nx
    fox=f(*ox)
    err=norm(fox)
    iter=iter+1
    print("Iteration:int:err={}:{:.7f}:{:.7f}".format(iter,interval,err))
print("[w1,w2,w3,w4,l1,l2]={}".format(ox))
print(fox)
print("No of iterations run is {}".format(iter))
print(sum(ox[:-2]))
print((fox[-1]+0.0576)**0.5)
#[w1,w2,w3,w4,l1,l2]=[ 0.05817187  0.59237059  0.6699484  -0.32049086 -0.07153799 -0.80002192]


1.0
0.11775902300885484
Iteration:int:err=1:11.0308376:2.2528458
Iteration:int:err=2:5.7874837:0.5285529
Iteration:int:err=3:2.8133350:0.0939220
Iteration:int:err=4:0.6809800:0.0244603
Iteration:int:err=5:0.2824860:0.0053266
Iteration:int:err=6:0.0599865:0.0002381
Iteration:int:err=7:0.0034371:0.0000006
Iteration:int:err=8:0.0000072:0.0000000
Iteration:int:err=9:0.0000000:0.0000000
[w1,w2,w3,w4,l1,l2]=[ 0.05817187  0.59237059  0.6699484  -0.32049086 -0.07153799 -0.80002192]
[1.73472348e-18 6.93889390e-18 0.00000000e+00 0.00000000e+00
 0.00000000e+00 6.93889390e-18]
No of iterations run is 9
1.0
0.24000000000000002


$\left[\begin{array}{c} w_1\\w_2\\w_3\\w_4\\\lambda_1\\\lambda_2\end{array}\right]=\left[\begin{array}{c} 0.05817187\\0.59237059\\0.6699484\\-0.32049086\\-0.07153799\\-0.80002192\end{array}\right]$  
$H(v,\lambda_1,\lambda_2)=\left[\begin{array}{cccc}
0.125\lambda_2& -0.015625\lambda_2&0.01875\lambda_2&0\\
-0.015625\lambda_2&0.125\lambda_2&-0.01875\lambda_2&0\\
0.01875\lambda_2&-0.01875\lambda_2&0.18\lambda_2&0\\
0&0&0&0.045\lambda_2\\
\end{array}\right]$  


In [23]:
import numpy as np
a=np.array([[0.125     ,-0.015625,0.01875 ,0],
            [-0.015625 ,0.125    ,-0.01875,0],
            [0.01875   ,-0.01875 ,0.18    ,0],
            [0         ,0        ,0       ,0.045]])
da=np.linalg.det(a)
print(da)
print(da*-0.80002192)
print(-0.80002192*a)
da=np.linalg.det(-0.80002192*a)
print(da)


0.00012112426757812485
-9.690206910644519e-05
[[-0.10000274  0.01250034 -0.01500041 -0.        ]
 [ 0.01250034 -0.10000274  0.01500041 -0.        ]
 [-0.01500041  0.01500041 -0.14400395 -0.        ]
 [-0.         -0.         -0.         -0.03600099]]
4.9617937753486526e-05



3) Use Newton's method to find the yield of a five year semiannual coupon bond with 3.375% coupon rate and price 100 1/32 What are the duration and convexity of the bond?

In [34]:
import numpy as np
import math
M=5
Freq=2
C=0.03375
B=100+1/32
F=100

t=np.arange(M,0,-1/Freq)[::-1]
print(t)
cf=np.full((len(t)),C*F/Freq)
cf[-1]+=F
print(cf)

f=lambda y:sum([cf[i]*math.exp(-y*t[i]) for i in range(len(t))])-B
f1=lambda y:sum([-cf[i]*t[i]*math.exp(-y*t[i]) for i in range(len(t))])

oy=0.035
foy=f(oy)

int_tol=10**(-6)
tol_app=10**(-9)
interval=abs(oy)
iters=0
while abs(foy)>tol_app or interval>int_tol:
    ny=oy-foy/f1(oy)
    interval=abs(ny-oy)
    oy=ny
    foy=f(oy)
    iters+=1
print("Yield:Price={:.6f}:{:.6f}".format(oy,foy+B))
print("Number of iterations is {}".format(iters))

print("Duration of the Bond is {:.4f}".format(-f1(oy)/B))
print("Convexity of the Bond is {:.4f}".format(sum([cf[i]*t[i]**2*math.exp(-oy*t[i]) for i in range(len(t))])/B))


[0.5 1.  1.5 2.  2.5 3.  3.5 4.  4.5 5. ]
[  1.6875   1.6875   1.6875   1.6875   1.6875   1.6875   1.6875   1.6875
   1.6875 101.6875]
Yield:Price=0.033401:100.031250
Number of iterations is 3
Duration of the Bond is 4.6427
Convexity of the Bond is 22.5731


4) Recall that finding the implied volatility from the given price of a call option is equivalent to solving the nonlinear problem f (x) = 0  
$f(x) = Se^{-qT}N(d_1(x)) - Ke^{-rT}N(d_2(x)) - C$  
and $d_1(x) =\frac{ln\left(\frac{S}{K}\right) + (r-q+\frac{x^2}{2})T}{x\sqrt{T}}\,\,\,d_2(x) =\frac{ln\left(\frac{S}{K}\right) + (r-q-\frac{x^2}{2})T}{x\sqrt{T}}$  
(i) Show that $lim_{x\rightarrow\infty}d_1(x) = \infty$ and $lim_{x\rightarrow\infty}d_2(x) = -\infty$ , and conclude that  
$lim_{x\rightarrow\infty}f(x)=Se^{-qT}-C$  

$Lt_{x\rightarrow\infty}d_1(x)\\
\quad= Lt_{x\rightarrow\infty}\frac{ln\left(\frac{S}{K}\right) + (r-q+\frac{x^2}{2})T}{x\sqrt{T}}\\
\quad= Lt_{x\rightarrow\infty}\frac{\frac{2x}{2}T}{\sqrt{T}}\\
\quad= Lt_{x\rightarrow\infty}x\sqrt{T}=\infty$   
$Lt_{x\rightarrow\infty}d_1(x)\\
\quad= Lt_{x\rightarrow\infty}-x\sqrt{T}=-\infty$

From properties of SNV $N(\infty)=1$ and $N(-\infty)=0$  
$Lt_{x\rightarrow\infty}f(x)\\
\quad= Lt_{x\rightarrow\infty}Se^{-qT}N(d_1(x)) - Ke^{-rT}N(d_2(x)) - C\\
\quad= Lt_{x\rightarrow\infty,d_1(x)\rightarrow\infty,d_2(x)\rightarrow-\infty}Se^{-qT}N(d_1(x)) - Ke^{-rT}N(d_2(x)) - C\\
\quad= Lt_{x\rightarrow\infty}Se^{-qT}*1 - Ke^{-rT}*0 - C\\
=Se^{-qT}- C$
  
  
  
(ii) Show that  
$lim_{x\rightarrow 0}d_1(x) = lim_{x\rightarrow 0}d_2(x) =$  
$-\infty\,\, if\,\, Se^{(r-q)T} < K\\0\,\,if\,\, Se^{(r-q)T}=K\\\infty\,\,if\,\, Se^{(r-q)T} >K$  

$d_1(x) =\frac{ln\left(\frac{S}{K}\right) + (r-q+\frac{x^2}{2})T}{x\sqrt{T}}\\
\quad=\frac{ln\left(\frac{S}{K}\right) + (r-q)T+\frac{x^2}{2}T}{x\sqrt{T}}\\
\quad=\frac{ln\left(\frac{S}{K}\right) + ln(e^{(r-q)T})+\frac{x^2}{2}T}{x\sqrt{T}}\\
\quad=\frac{ln\left(\frac{Se^{(r-q)T}}{K}\right) +\frac{x^2}{2}T}{x\sqrt{T}}$  

From the above forward price based definition of d_1(x) and d_2(x), we can prove the inequalities.




---
5) A three months at-the-money call on an underlying asset with spot price 30 paying dividends continuously at a 2% rate is worth \$2.5. Assume that the nsk free mterest rate is constant at 6\%.  
(i). Compute the implied volatility with six decimal digits accuracy, using the bisection method on the interval [0.0001, 1], the secant method wIth Imtlal guess 0.5, and Newton's method with initial guess 0.5.  
(ii) Let $\sigma_{imp}$ be the implied volatility previously computed using Newton's method. Use formula (5.77) to compute an approximate value $\sigma_{imp,approx}$ for the ImplIed volatility, and compute the relative error  
$\frac{|\sigma_{imp,approx} - \sigma_{imp}|}{\sigma_{imp}}$  


*** Bisection Method***

In [42]:
from scipy.stats import norm
S=30
q=0.02
T=3/12
K=30
C=2.5
r=0.06
print("S,K,r,q,v,T={},{},{},{},{},{}".format(S,K,r,q,0,T))

d1=lambda S,K,r,q,v,T:(math.log(S/K)+(r-q+v**2/2)*T)/v/T**0.5
d2=lambda S,K,r,q,v,T:(math.log(S/K)+(r-q-v**2/2)*T)/v/T**0.5
V=lambda S,K,r,q,v,T:S*math.exp(-q*T)*norm.cdf(d1(S,K,r,q,v,T))-K*math.exp(-r*T)*norm.cdf(d2(S,K,r,q,v,T))
f=lambda v:V(S,K,r,q,v,T)-C

#Not sure if is continoulsy monotonous function in the domain.
#For bisection method to work this has to be true!
lv=0.0001
rv=1
print("vol:N(d1):N(d2)=={:.2%}:{:.6f}:{:.6f}".format(lv,norm.cdf(d1(S,K,r,q,lv,T)),norm.cdf(d2(S,K,r,q,lv,T))))
print("vol:N(d1):N(d2)=={:.2%}:{:.6f}:{:.6f}".format(rv,norm.cdf(d1(S,K,r,q,rv,T)),norm.cdf(d2(S,K,r,q,rv,T))))
lf=f(lv)
rf=f(rv)
print(lf)
print(rf)
tol=10**(-9)
v_err=10**(-6)
midf=lf
mid=lv
iters=0
while abs(midf)>tol or rv-lv>v_err:
    mid=(lv+rv)/2
    midf=f(mid)
    if midf<0 and lf<0:
        lv=mid
        lf=midf
    else:
        rv=mid
        rf=midf
    iters+=1
    print("Iter:f:V={}:{:.6f}:{:.6f}".format(iters,midf,midf+C))
print("Implied Volatility is {:.2%}".format(mid))


S,K,r,q,v,T=30,30,0.06,0.02,0,0.25
vol:N(d1):N(d2)==0.01%:1.000000:1.000000
vol:N(d1):N(d2)==100.00%:0.606420:0.409046
-2.2029838123114125
3.5131806933015124
Iter:f:V=1:0.605789:3.105789
Iter:f:V=2:-0.866119:1.633881
Iter:f:V=3:-0.129479:2.370521
Iter:f:V=4:0.238413:2.738413
Iter:f:V=5:0.054523:2.554523
Iter:f:V=6:-0.037466:2.462534
Iter:f:V=7:0.008532:2.508532
Iter:f:V=8:-0.014466:2.485534
Iter:f:V=9:-0.002967:2.497033
Iter:f:V=10:0.002782:2.502782
Iter:f:V=11:-0.000092:2.499908
Iter:f:V=12:0.001345:2.501345
Iter:f:V=13:0.000626:2.500626
Iter:f:V=14:0.000267:2.500267
Iter:f:V=15:0.000087:2.500087
Iter:f:V=16:-0.000002:2.499998
Iter:f:V=17:0.000042:2.500042
Iter:f:V=18:0.000020:2.500020
Iter:f:V=19:0.000009:2.500009
Iter:f:V=20:0.000003:2.500003
Iter:f:V=21:0.000000:2.500000
Iter:f:V=22:-0.000001:2.499999
Iter:f:V=23:-0.000000:2.500000
Iter:f:V=24:0.000000:2.500000
Iter:f:V=25:-0.000000:2.500000
Iter:f:V=26:-0.000000:2.500000
Iter:f:V=27:-0.000000:2.500000
Iter:f:V=28:0.000000:2.500000

*** Secant Method***

In [44]:
from scipy.stats import norm
S=30
q=0.02
T=3/12
K=30
C=2.5
r=0.06
print("S,K,r,q,v,T={},{},{},{},{},{}".format(S,K,r,q,0,T))

d1=lambda S,K,r,q,v,T:(math.log(S/K)+(r-q+v**2/2)*T)/v/T**0.5
d2=lambda S,K,r,q,v,T:(math.log(S/K)+(r-q-v**2/2)*T)/v/T**0.5
V=lambda S,K,r,q,v,T:S*math.exp(-q*T)*norm.cdf(d1(S,K,r,q,v,T))-K*math.exp(-r*T)*norm.cdf(d2(S,K,r,q,v,T))
f=lambda v:V(S,K,r,q,v,T)-C
f1=lambda ov,oov,fov,foov:(foov-fov)/(oov-ov)
#Not sure if is continoulsy monotonous function in the domain.
#For bisection method to work this has to be true!
ov=0.5
oov=0.4
tol=10**(-9)
v_err=10**(-6)
fov=f(ov)
foov=f(oov)
iters=0
while abs(fov)>tol or abs(fov-foov)>v_err:
    nv=ov-fov/f1(ov,oov,fov,foov)
    oov=ov
    foov=fov
    ov=nv
    fov=f(nv)
    iters+=1
    print("Iter:f:V={}:{:.6f}:{:.6f}".format(iters,fov,fov+C))
print("Implied Volatility is {:.2%}".format(ov))


S,K,r,q,v,T=30,30,0.06,0.02,0,0.25
Iter:f:V=1:-0.000020:2.499980
Iter:f:V=2:0.000000:2.500000
Iter:f:V=3:0.000000:2.500000
Implied Volatility is 39.70%


**Newton's Method**  
$f(\sigma)=Se^{-qT}N(d_1)-Ke^{-rT}N(d_2)-C$  
$f'(\sigma)=Se^{-qT}N'(d_1)\left(\frac{dd_1}{d\sigma}-\frac{dd_2}{d\sigma}\right)$  
$\frac{dd_1}{d\sigma}-\frac{dd_2}{d\sigma}=\sqrt{T}$  
$f'(\sigma)=Se^{-qT}\sqrt{T}N'(d_1)$  
$f'(\sigma)=Se^{-qT}\sqrt{\frac{T}{2\pi}}e^{-\frac{d_1^2}{2}}$  

In [46]:
from scipy.stats import norm
S=30
q=0.02
T=3/12
K=30
C=2.5
r=0.06
print("S,K,r,q,v,T={},{},{},{},{},{}".format(S,K,r,q,0,T))

d1=lambda S,K,r,q,v,T:(math.log(S/K)+(r-q+v**2/2)*T)/v/T**0.5
d2=lambda S,K,r,q,v,T:(math.log(S/K)+(r-q-v**2/2)*T)/v/T**0.5
V=lambda S,K,r,q,v,T:S*math.exp(-q*T)*norm.cdf(d1(S,K,r,q,v,T))-K*math.exp(-r*T)*norm.cdf(d2(S,K,r,q,v,T))
f=lambda v:V(S,K,r,q,v,T)-C
f1=lambda v:S*math.exp(-q*T)*(T/2/math.pi)**0.5*math.exp(-d1(S,K,r,q,v,T)**2/2)

#Not sure if is continoulsy monotonous function in the domain.
#For bisection method to work this has to be true!
ov=0.5
tol=10**(-9)
v_err=10**(-6)
fov=f(ov)
interval=ov
iters=0
while abs(fov)>tol or interval>v_err:
    nv=ov-fov/f1(ov)
    interval=abs(nv-ov)
    ov=nv
    fov=f(ov)
    iters+=1
    print("Iter:f:V={}:{:.6f}:{:.6f}".format(iters,fov,fov+C))
print("Implied Volatility is {:.2%}".format(ov))


S,K,r,q,v,T=30,30,0.06,0.02,0,0.25
Iter:f:V=1:-0.000783:2.499217
Iter:f:V=2:-0.000000:2.500000
Iter:f:V=3:0.000000:2.500000
Implied Volatility is 39.70%


---
***Exercise 6***  
$F:R^3\rightarrow R^3$  
$F=\left[\begin{array}{c}x_1^3+2x_1x_2+x_3^2-x_2x_3+9\\
2x_1^2+2x_1x_2^2+x_2^3x_3^2-x_2^2x_3-2\\
x_1x_2x_3+x_1^3-x_3^2-x_1x_2^2-4\end{array}\right]$  

Solve F(x)=0 with starting points  
$x_0=\left[\begin{array}{c}1\\2\\3\end{array}\right]$  
$x_1=\left[\begin{array}{}2\\2\\2\end{array}\right]$

In [2]:
import numpy as np
f=lambda x1,x2,x3:np.array([x1**3  + 2*x1*x2    + x3**2         - x2*x3    + 9,
                           2*x1**2 + 2*x1*x2**2 + x2**3 * x3**2 - x2**2*x3 - 2,
                           x1*x2*x3+ x1**3      - x3**2         - x1*x2**2 -4])
h=10**(-6)
f1=lambda x1,x2,x3:np.c_[(f(x1+h,x2,x3)-f(x1-h,x2,x3))/2/h,(f(x1,x2+h,x3)-f(x1,x2-h,x3))/2/h,(f(x1,x2,x3+h)-f(x1,x2,x3-h))/2/h]

norm=lambda a:sum(a*a)**0.5
int_tol=10**(-6)
val_tol=10**(-9)

ov=np.array([2,2,2])
interval=norm(ov)
fov=f(*ov)
err=norm(fov)

iters=0
while err>val_tol or interval>int_tol:
    nv=ov-np.linalg.solve(f1(*ov),fov)
    interval=norm(nv-ov)
    ov=nv
    fov=f(*ov)
    err=norm(fov)
    iters=iters+1
    print("Iter:X:fov={},{},{}".format(iters,ov,err))

Iter:X:fov=1,[-0.4516129  -5.35483869 16.4677419 ],42141.60188112772
Iter:X:fov=2,[ 0.21571499 -5.66418295  6.75906943],8507.700856323045
Iter:X:fov=3,[ 0.28602268 -6.52533236  1.7962525 ],950.9346071091666
Iter:X:fov=4,[ -0.25842516 -12.48640201  -1.61432306],4904.1690441679375
Iter:X:fov=5,[ -0.08943886 -11.02414748  -1.10705907],1531.2003368904288
Iter:X:fov=6,[  0.72871735 -46.46472689   4.67296072],2197500.210997958
Iter:X:fov=7,[ -1.10342394 -88.17589205  -3.96511375],10764895.732919607
Iter:X:fov=8,[ -4.72168514  90.39763372 -14.02412905],145323220.81208473
Iter:X:fov=9,[ 60.64579881 781.02406327 153.71653876],11257285356094.043
Iter:X:fov=10,[ 42.27055528 536.01113678 149.19144693],3427732628573.2534
Iter:X:fov=11,[ 30.58979958 370.12129031 143.8557824 ],1049257693722.1229
Iter:X:fov=12,[ 22.57223142 260.76282292 135.68514376],326432529665.24884
Iter:X:fov=13,[ 16.41148752 189.53299212 123.4386729 ],103739301465.21849
Iter:X:fov=14,[ 11.38876764 141.72835019 108.42135895],33463

[1,2,3]$\rightarrow$[-1.69055076  1.98310724 -0.88455808]  
[2,2,2]$\rightarrow$[-1.  3.  1.]

(i) Use bootstrapping to obtain a zero rate curve from the following prices of Treasury instruments with semiannual coupon payments:  

Security|Coupon|Price
---|---|---
3-Month T-bill|0|98.7 
6-Month T-bill|0|97.5
2-Year T-bond|4.875|100$\frac{5}{32}$ 
3-Year T-bond|4.875|100$\frac{5}{32}$
5-Year T-bond|4.625|99$\frac{22}{32}$
10-Year T-bond|4.875|101$\frac{22}{32}$

Assume that interest is continuously compounded.


-1.1690087971667396
[-24.382747800708316]
0.052055904169554676
[0.0520559]
Iteration:Yield:err=1:0.0520559:0.0070340
[-24.67675848967446]
0.05234094803788833
[0.05234095]
Iteration:Yield:err=2:0.0523409:0.0000003
[-24.67500006265435]
0.052340958194622006
[0.05234096]
Iteration:Yield:err=3:0.0523410:0.0000000
Backfilling zero rate for 0.25 years with 5.2341%
{0.25: 0.052340958194622006}
-2.377057549928594
[-47.5614712250357]
0.05002136206675324
[0.05002136]
Iteration:Yield:err=1:0.0500214:0.0299495
[-48.76497473831217]
0.050635521651271775
[0.05063552]
Iteration:Yield:err=2:0.0506355:0.0000046
[-48.750002298984434]
0.050635615968577435
[0.05063562]
Iteration:Yield:err=3:0.0506356:0.0000000
Backfilling zero rate for 0.50 years with 5.0636%
{0.25: 0.052340958194622006, 0.5: 0.050635615968577435}
nan
[nan, nan, -167.73746303685152]
nan
[0.05063562        nan        nan        nan]
Iteration:Yield:err=1:nan:nan
Backfilling zero rate for 1.00 years with nan%
Backfilling zero rate for 1.50 ye

In [7]:
a=np.full((4),0.)
a[2]=np.nan
print(np.isnan(a))
print(np.isnan(a))
a=[[1,2],
  [2,3],
  [3,4]]
print(len(a))

[False False  True False]
[False False  True False]
3
